

<!DOCTYPE html>
<html >
  <head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width,initial-scale=1">
      <title>Core Classes</title>
    
      <link rel="stylesheet" href="../_static/pygments.css">
      <link rel="stylesheet" href="../_static/theme.css">
      <link rel="stylesheet" href="../_static/css/custom.css">
      
      <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>

      <!-- sphinx script_files -->
        <script type="text/javascript" src="../_static/jquery.js"></script>
        <script type="text/javascript" src="../_static/underscore.js"></script>
        <script type="text/javascript" src="../_static/doctools.js"></script>
        <script type="text/javascript" src="../_static/language_data.js"></script>

      
      <script src="../_static/theme-vendors.js"></script>
      <script src="../_static/theme.js" defer></script>
    
  <link rel="index" title="Index" href="../genindex.html" />
  <link rel="search" title="Search" href="../search.html" />
  <link rel="next" title="Reports" href="reports.html" />
  <link rel="prev" title="DNA Chisel API reference" href="ref.html" /> 
  </head>

  <body><div id="app" class="theme-container" :class="pageClasses"><navbar @toggle-sidebar="toggleSidebar">
  <router-link to="../index.html" class="home-link">
    
      <img class="logo" src="../_static/html_logo.png" alt="doit logo"/>
    
  </router-link>

  <div class="links">
    <navlinks class="can-hide">

  
    <div class="nav-item">
      <a href="../index.html#more-biology-software"
         class="nav-link ">
         DNA Chisel documentation
      </a>
    </div>
  
    <div class="nav-item">
      <a href="ref.html#dna-chisel-api-reference"
         class="nav-link  router-link-active">
         DNA Chisel API reference
      </a>
    </div>
  
    <div class="nav-item">
      <a href="../index.html#more-biology-software"
         class="nav-link ">
         Examples
      </a>
    </div>
  



    </navlinks>
  </div>
</navbar>

      
      <div class="sidebar-mask" @click="toggleSidebar(false)">
      </div>
        <sidebar @toggle-sidebar="toggleSidebar">
          
          <navlinks>
            

  
    <div class="nav-item">
      <a href="../index.html#more-biology-software"
         class="nav-link ">
         DNA Chisel documentation
      </a>
    </div>
  
    <div class="nav-item">
      <a href="ref.html#dna-chisel-api-reference"
         class="nav-link  router-link-active">
         DNA Chisel API reference
      </a>
    </div>
  
    <div class="nav-item">
      <a href="../index.html#more-biology-software"
         class="nav-link ">
         Examples
      </a>
    </div>
  



            
          </navlinks><div id="searchbox" class="searchbox" role="search">
  <div class="caption"><span class="caption-text">Quick search</span>
    <div class="searchformwrapper">
      <form class="search" action="../search.html" method="get">
        <input type="text" name="q" />
        <input type="submit" value="Search" />
        <input type="hidden" name="check_keywords" value="yes" />
        <input type="hidden" name="area" value="default" />
      </form>
    </div>
  </div>
</div><div class="sidebar-links" role="navigation" aria-label="main navigation">
  
    <div class="sidebar-group">
      <p class="caption">
        <span class="caption-text"><a href="../index.html#more-biology-software">DNA Chisel documentation</a></span>
      </p>
      <ul class="">
        
      </ul>
    </div>
  
    <div class="sidebar-group">
      <p class="caption">
        <span class="caption-text"><a href="ref.html#dna-chisel-api-reference">DNA Chisel API reference</a></span>
      </p>
      <ul class="current">
        
          <li class="toctree-l1 current"><a href="#" class="reference internal current">Core Classes</a>

            
              <ul>
                
                  <li class="toctree-l2"><a href="#dnaoptimizationproblem" class="reference internal">DnaOptimizationProblem</a></li>
                
                  <li class="toctree-l2"><a href="#circulardnaoptimizationproblem" class="reference internal">CircularDnaOptimizationProblem</a></li>
                
                  <li class="toctree-l2"><a href="#specification" class="reference internal">Specification</a></li>
                
                  <li class="toctree-l2"><a href="#specevaluation" class="reference internal">SpecEvaluation</a></li>
                
                  <li class="toctree-l2"><a href="#module-dnachisel.Location" class="reference internal">Location</a></li>
                
                  <li class="toctree-l2"><a href="#module-dnachisel.MutationSpace" class="reference internal">MutationSpace</a></li>
                
                  <li class="toctree-l2"><a href="#sequencepattern" class="reference internal">SequencePattern</a></li>
                
              </ul>
            
          </li>

        
          <li class="toctree-l1 "><a href="reports.html" class="reference internal ">Reports</a>

            
          </li>

        
          <li class="toctree-l1 "><a href="biotools.html" class="reference internal ">Biotools</a>

            
          </li>

        
          <li class="toctree-l1 "><a href="builtin_specifications.html" class="reference internal ">Built-in Specifications</a>

            
          </li>

        
      </ul>
    </div>
  
    <div class="sidebar-group">
      <p class="caption">
        <span class="caption-text"><a href="../index.html#more-biology-software">Examples</a></span>
      </p>
      <ul class="">
        
          <li class="toctree-l1 "><a href="../examples.html" class="reference internal ">Examples</a>

            
          </li>

        
      </ul>
    </div>
  
</div>
        </sidebar>

      <page>
          <div class="body-header" role="navigation" aria-label="navigation">
  
  <ul class="breadcrumbs">
    <li><a href="../index.html">Docs</a> &raquo;</li>
    
      <li><a href="ref.html">DNA Chisel API reference</a> &raquo;</li>
    
    <li>Core Classes</li>
  </ul>
  

  <ul class="page-nav">
  <li class="prev">
    <a href="ref.html"
       title="previous chapter">← DNA Chisel API reference</a>
  </li>
  <li class="next">
    <a href="reports.html"
       title="next chapter">Reports →</a>
  </li>
</ul>
  
</div>
<hr>
          <div class="content" role="main">
            
  <div class="section" id="core-classes">
<h1><a class="toc-backref" href="#id1">Core Classes</a><a class="headerlink" href="#core-classes" title="Permalink to this headline">¶</a></h1>
<div class="contents topic" id="contents">
<p class="topic-title first">Contents</p>
<ul class="simple">
<li><p><a class="reference internal" href="#core-classes" id="id1">Core Classes</a></p>
<ul>
<li><p><a class="reference internal" href="#dnaoptimizationproblem" id="id2">DnaOptimizationProblem</a></p></li>
<li><p><a class="reference internal" href="#circulardnaoptimizationproblem" id="id3">CircularDnaOptimizationProblem</a></p></li>
<li><p><a class="reference internal" href="#specification" id="id4">Specification</a></p></li>
<li><p><a class="reference internal" href="#specevaluation" id="id5">SpecEvaluation</a></p></li>
<li><p><a class="reference internal" href="#module-dnachisel.Location" id="id6">Location</a></p></li>
<li><p><a class="reference internal" href="#module-dnachisel.MutationSpace" id="id7">MutationSpace</a></p></li>
<li><p><a class="reference internal" href="#sequencepattern" id="id8">SequencePattern</a></p></li>
</ul>
</li>
</ul>
</div>
<div class="section" id="dnaoptimizationproblem">
<h2><a class="toc-backref" href="#id2">DnaOptimizationProblem</a><a class="headerlink" href="#dnaoptimizationproblem" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="dnachisel.DnaOptimizationProblem">
<em class="property">class </em><code class="sig-prename descclassname">dnachisel.</code><code class="sig-name descname">DnaOptimizationProblem</code><span class="sig-paren">(</span><em class="sig-param">sequence</em>, <em class="sig-param">constraints=None</em>, <em class="sig-param">objectives=None</em>, <em class="sig-param">logger='bar'</em>, <em class="sig-param">mutation_space=None</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dnachisel/DnaOptimizationProblem.html#DnaOptimizationProblem"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dnachisel.DnaOptimizationProblem" title="Permalink to this definition">¶</a></dt>
<dd><p>Problem specifications: sequence, constraints, optimization objectives.</p>
<p>The original constraints, objectives, and original sequence of the problem
are stored in the DNA Canvas. This class also has methods to display
reports on the constraints and objectives, as well as solving the
constraints and objectives.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>sequence</strong></p>
<blockquote>
<div><p>A string of ATGC characters (they must be upper case!), e.g. “ATTGTGTA”</p>
</div></blockquote>
<p><strong>constraints</strong></p>
<blockquote>
<div><p>A list of objects of type <code class="docutils literal notranslate"><span class="pre">Specification</span></code>.</p>
</div></blockquote>
<p><strong>objectives</strong></p>
<blockquote>
<div><p>A list of objects of type <code class="docutils literal notranslate"><span class="pre">Specification</span></code> specifying what must be
optimized in the problem. Note that each objective has a float <code class="docutils literal notranslate"><span class="pre">boost</span></code>
parameter. The larger the boost, the more the objective is taken into
account during the optimization.</p>
</div></blockquote>
<p><strong>logger</strong></p>
<blockquote>
<div><p>Either None for no logger, ‘bar’ for a tqdm progress bar logger, or
any ProgLog progress bar logger.</p>
</div></blockquote>
<p><strong>mutations_space</strong></p>
<blockquote>
<div><p>A MutationSpace indicating the possible mutations. In most case the
mutation space will be left to None and computed at problem
initialization (which can be slightly compute-intensive), however some
core DNA Chisel methods will create optimization problems with a provided
mutation_space to save computing time.</p>
</div></blockquote>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>The dictionnary <code class="docutils literal notranslate"><span class="pre">self.possible_mutations</span></code> is of the form
<code class="docutils literal notranslate"><span class="pre">{location1</span> <span class="pre">:</span> <span class="pre">list1,</span> <span class="pre">location2:</span> <span class="pre">list2...}</span></code>
where <code class="docutils literal notranslate"><span class="pre">location</span></code> is either a single index (e.g. 10) indicating the
position of a nucleotide to be muted, or a couple <code class="docutils literal notranslate"><span class="pre">(start,</span> <span class="pre">end)</span></code>
indicating a whole segment whose sub-sequence should be replaced.
The <code class="docutils literal notranslate"><span class="pre">list</span></code> s are lists of possible sequences to replace each location,
e.g. for the mutation of a whole codon <code class="docutils literal notranslate"><span class="pre">(3,6):</span> <span class="pre">[&quot;ATT&quot;,</span> <span class="pre">&quot;ACT&quot;,</span> <span class="pre">&quot;AGT&quot;]</span></code>.</p>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">dnachisel</span> <span class="k">import</span> <span class="o">*</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">problem</span> <span class="o">=</span> <span class="n">DnaOptimizationProblem</span><span class="p">(</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">sequence</span> <span class="o">=</span> <span class="s2">&quot;ATGCGTGTGTGC...&quot;</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">constraints</span> <span class="o">=</span> <span class="p">[</span><span class="n">constraint1</span><span class="p">,</span> <span class="n">constraint2</span><span class="p">,</span> <span class="o">...</span><span class="p">],</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">objectives</span> <span class="o">=</span> <span class="p">[</span><span class="n">objective1</span><span class="p">,</span> <span class="n">objective2</span><span class="p">,</span> <span class="o">...</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">problem</span><span class="o">.</span><span class="n">resolve_constraints</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">problem</span><span class="o">.</span><span class="n">optimize</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">canvas</span><span class="o">.</span><span class="n">constraints_text_summary</span><span class="p">())</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">canvas</span><span class="o">.</span><span class="n">objectives_text_summary</span><span class="p">())</span>
</pre></div>
</div>
<p class="rubric">Attributes</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 39%" />
<col style="width: 61%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><strong>sequence</strong></p></td>
<td><p>The sequence</p></td>
</tr>
<tr class="row-even"><td><p><strong>constraints</strong></p></td>
<td><p>The list of constraints</p></td>
</tr>
<tr class="row-odd"><td><p><strong>objectives</strong></p></td>
<td><p>The list of objectives</p></td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="dnachisel.DnaOptimizationProblem.all_constraints_pass">
<code class="sig-name descname">all_constraints_pass</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dnachisel/DnaOptimizationProblem.html#DnaOptimizationProblem.all_constraints_pass"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dnachisel.DnaOptimizationProblem.all_constraints_pass" title="Permalink to this definition">¶</a></dt>
<dd><p>Return True iff the current problem sequence passes all constraints.</p>
</dd></dl>

<dl class="method">
<dt id="dnachisel.DnaOptimizationProblem.constraints_evaluations">
<code class="sig-name descname">constraints_evaluations</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dnachisel/DnaOptimizationProblem.html#DnaOptimizationProblem.constraints_evaluations"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dnachisel.DnaOptimizationProblem.constraints_evaluations" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a list of the evaluations of each constraint of the canvas.</p>
</dd></dl>

<dl class="method">
<dt id="dnachisel.DnaOptimizationProblem.from_record">
<em class="property">static </em><code class="sig-name descname">from_record</code><span class="sig-paren">(</span><em class="sig-param">record</em>, <em class="sig-param">specifications_dict='default'</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dnachisel/DnaOptimizationProblem.html#DnaOptimizationProblem.from_record"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dnachisel.DnaOptimizationProblem.from_record" title="Permalink to this definition">¶</a></dt>
<dd><p>TODO: docs</p>
</dd></dl>

<dl class="method">
<dt id="dnachisel.DnaOptimizationProblem.initialize">
<code class="sig-name descname">initialize</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dnachisel/DnaOptimizationProblem.html#DnaOptimizationProblem.initialize"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dnachisel.DnaOptimizationProblem.initialize" title="Permalink to this definition">¶</a></dt>
<dd><p>Variables initialization before solving.</p>
</dd></dl>

<dl class="method">
<dt id="dnachisel.DnaOptimizationProblem.objectives_evaluations">
<code class="sig-name descname">objectives_evaluations</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dnachisel/DnaOptimizationProblem.html#DnaOptimizationProblem.objectives_evaluations"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dnachisel.DnaOptimizationProblem.objectives_evaluations" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a list of the evaluation of each objective of the canvas</p>
</dd></dl>

<dl class="method">
<dt id="dnachisel.DnaOptimizationProblem.optimize">
<code class="sig-name descname">optimize</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dnachisel/DnaOptimizationProblem.html#DnaOptimizationProblem.optimize"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dnachisel.DnaOptimizationProblem.optimize" title="Permalink to this definition">¶</a></dt>
<dd><p>Maximize the objective via local, targeted mutations.</p>
</dd></dl>

<dl class="method">
<dt id="dnachisel.DnaOptimizationProblem.optimize_by_exhaustive_search">
<code class="sig-name descname">optimize_by_exhaustive_search</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dnachisel/DnaOptimizationProblem.html#DnaOptimizationProblem.optimize_by_exhaustive_search"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dnachisel.DnaOptimizationProblem.optimize_by_exhaustive_search" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="dnachisel.DnaOptimizationProblem.optimize_by_random_mutations">
<code class="sig-name descname">optimize_by_random_mutations</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dnachisel/DnaOptimizationProblem.html#DnaOptimizationProblem.optimize_by_random_mutations"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dnachisel.DnaOptimizationProblem.optimize_by_random_mutations" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="dnachisel.DnaOptimizationProblem.optimize_with_report">
<code class="sig-name descname">optimize_with_report</code><span class="sig-paren">(</span><em class="sig-param">target</em>, <em class="sig-param">project_name='My project'</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dnachisel/DnaOptimizationProblem.html#DnaOptimizationProblem.optimize_with_report"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dnachisel.DnaOptimizationProblem.optimize_with_report" title="Permalink to this definition">¶</a></dt>
<dd><p>Resolve constraints, optimize objectives, write a multi-file report.</p>
<p>The report’s content may vary depending on the optimization’s success.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>target</strong></p>
<blockquote>
<div><p>Either a path to a folder that will containt the report, or a path to
a zip archive, or “&#64;memory” to return raw data of a zip archive
containing the report.</p>
</div></blockquote>
<p><strong>project_name</strong></p>
<blockquote>
<div><p>Project name to write on PDF reports</p>
</div></blockquote>
<p><strong>**solver_parameters</strong></p>
<blockquote>
<div><p>Extra keyword arguments passed to <code class="docutils literal notranslate"><span class="pre">problem.resolve_constraints()</span></code></p>
</div></blockquote>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>(success, message, zip_data)</p>
<blockquote>
<div><p>Triplet where success is True/False, message is a one-line string
summary indication whether some clash was found, or some solution, or
maybe no solution was found because the random searches were too short</p>
</div></blockquote>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="dnachisel.DnaOptimizationProblem.resolve_constraint">
<code class="sig-name descname">resolve_constraint</code><span class="sig-paren">(</span><em class="sig-param">constraint</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dnachisel/DnaOptimizationProblem.html#DnaOptimizationProblem.resolve_constraint"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dnachisel.DnaOptimizationProblem.resolve_constraint" title="Permalink to this definition">¶</a></dt>
<dd><p>Resolve a constraint through successive localizations.</p>
</dd></dl>

<dl class="method">
<dt id="dnachisel.DnaOptimizationProblem.resolve_constraints">
<code class="sig-name descname">resolve_constraints</code><span class="sig-paren">(</span><em class="sig-param">final_check=True</em>, <em class="sig-param">cst_filter=None</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dnachisel/DnaOptimizationProblem.html#DnaOptimizationProblem.resolve_constraints"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dnachisel.DnaOptimizationProblem.resolve_constraints" title="Permalink to this definition">¶</a></dt>
<dd><p>Solve a particular constraint using local, targeted searches.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>constraint</strong></p>
<blockquote>
<div><p>The <code class="docutils literal notranslate"><span class="pre">Specification</span></code> object for which the sequence should be solved</p>
</div></blockquote>
<p><strong>final_check</strong></p>
<blockquote>
<div><p>If True, a final check of that all constraints pass will be run at
the end of the process, when constraints have been resolved one by
one, to check that the solving of one constraint didn’t undo the
solving of another.</p>
</div></blockquote>
<p><strong>cst_filter</strong></p>
<blockquote>
<div><p>An optional filter to only resolve a subset function (constraint =&gt; True/False)</p>
</div></blockquote>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="dnachisel.DnaOptimizationProblem.resolve_constraints_by_exhaustive_search">
<code class="sig-name descname">resolve_constraints_by_exhaustive_search</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dnachisel/DnaOptimizationProblem.html#DnaOptimizationProblem.resolve_constraints_by_exhaustive_search"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dnachisel.DnaOptimizationProblem.resolve_constraints_by_exhaustive_search" title="Permalink to this definition">¶</a></dt>
<dd><p>Solve all constraints by exploring the whole search space.</p>
<p>This method iterates over <code class="docutils literal notranslate"><span class="pre">self.iter_mutations_space()</span></code> (space of
all sequences that could be reached through successive mutations) and
stops when it finds a sequence which meets all the constraints of the
canvas.</p>
</dd></dl>

<dl class="method">
<dt id="dnachisel.DnaOptimizationProblem.resolve_constraints_by_random_mutations">
<code class="sig-name descname">resolve_constraints_by_random_mutations</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dnachisel/DnaOptimizationProblem.html#DnaOptimizationProblem.resolve_constraints_by_random_mutations"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dnachisel.DnaOptimizationProblem.resolve_constraints_by_random_mutations" title="Permalink to this definition">¶</a></dt>
<dd><p>Solve all constraints by successive sets of random mutations.</p>
<p>This method modifies the canvas sequence by applying a number
<code class="docutils literal notranslate"><span class="pre">n_mutations</span></code> of random mutations. The constraints are then evaluated
on the new sequence. If all constraints pass, the new sequence becomes
the canvas’s new sequence.
If not all constraints pass, the sum of all scores from failing
constraints is considered. If this score is superior to the score of
the previous sequence, the new sequence becomes the canvas’s new
sequence.</p>
<p>This operation is repeated <cite>max_iter</cite> times at most, after which
a <code class="docutils literal notranslate"><span class="pre">NoSolutionError</span></code> is thrown.</p>
</dd></dl>

<dl class="method">
<dt id="dnachisel.DnaOptimizationProblem.resolve_constraints_locally">
<code class="sig-name descname">resolve_constraints_locally</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dnachisel/DnaOptimizationProblem.html#DnaOptimizationProblem.resolve_constraints_locally"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dnachisel.DnaOptimizationProblem.resolve_constraints_locally" title="Permalink to this definition">¶</a></dt>
<dd><p>Orient the local search towards a stochastic or exhaustive search.</p>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="circulardnaoptimizationproblem">
<h2><a class="toc-backref" href="#id3">CircularDnaOptimizationProblem</a><a class="headerlink" href="#circulardnaoptimizationproblem" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="dnachisel.CircularDnaOptimizationProblem">
<em class="property">class </em><code class="sig-prename descclassname">dnachisel.</code><code class="sig-name descname">CircularDnaOptimizationProblem</code><span class="sig-paren">(</span><em class="sig-param">sequence</em>, <em class="sig-param">constraints=None</em>, <em class="sig-param">objectives=None</em>, <em class="sig-param">logger='bar'</em>, <em class="sig-param">mutation_space=None</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dnachisel/CircularDnaOptimizationProblem.html#CircularDnaOptimizationProblem"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dnachisel.CircularDnaOptimizationProblem" title="Permalink to this definition">¶</a></dt>
<dd><p>Class for solving circular DNA optimization problems.</p>
<p>The high-level interface is the same as for DnaOptimizationProblem:
initialization, resolve_constraints(), and optimize() work the same way.</p>
<dl class="method">
<dt id="dnachisel.CircularDnaOptimizationProblem.constraints_evaluations">
<code class="sig-name descname">constraints_evaluations</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dnachisel/CircularDnaOptimizationProblem.html#CircularDnaOptimizationProblem.constraints_evaluations"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dnachisel.CircularDnaOptimizationProblem.constraints_evaluations" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a list of the evaluations of each constraint of the canvas.</p>
</dd></dl>

<dl class="method">
<dt id="dnachisel.CircularDnaOptimizationProblem.objectives_evaluations">
<code class="sig-name descname">objectives_evaluations</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dnachisel/CircularDnaOptimizationProblem.html#CircularDnaOptimizationProblem.objectives_evaluations"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dnachisel.CircularDnaOptimizationProblem.objectives_evaluations" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a list of the evaluation of each objective of the canvas</p>
</dd></dl>

<dl class="method">
<dt id="dnachisel.CircularDnaOptimizationProblem.optimize">
<code class="sig-name descname">optimize</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dnachisel/CircularDnaOptimizationProblem.html#CircularDnaOptimizationProblem.optimize"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dnachisel.CircularDnaOptimizationProblem.optimize" title="Permalink to this definition">¶</a></dt>
<dd><p>Maximize the objective via local, targeted mutations.</p>
</dd></dl>

<dl class="method">
<dt id="dnachisel.CircularDnaOptimizationProblem.optimize_with_report">
<code class="sig-name descname">optimize_with_report</code><span class="sig-paren">(</span><em class="sig-param">target</em>, <em class="sig-param">project_name='My project'</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dnachisel/CircularDnaOptimizationProblem.html#CircularDnaOptimizationProblem.optimize_with_report"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dnachisel.CircularDnaOptimizationProblem.optimize_with_report" title="Permalink to this definition">¶</a></dt>
<dd><p>Resolve constraints, optimize objectives, write a multi-file report.</p>
<p>The report’s content may vary depending on the optimization’s success.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>target</strong></p>
<blockquote>
<div><p>Either a path to a folder that will containt the report, or a path to
a zip archive, or “&#64;memory” to return raw data of a zip archive
containing the report.</p>
</div></blockquote>
<p><strong>project_name</strong></p>
<blockquote>
<div><p>Project name to write on PDF reports</p>
</div></blockquote>
<p><strong>**solver_parameters</strong></p>
<blockquote>
<div><p>Extra keyword arguments passed to <code class="docutils literal notranslate"><span class="pre">problem.resolve_constraints()</span></code></p>
</div></blockquote>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>(success, message, zip_data)</p>
<blockquote>
<div><p>Triplet where success is True/False, message is a one-line string
summary indication whether some clash was found, or some solution, or
maybe no solution was found because the random searches were too short</p>
</div></blockquote>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="dnachisel.CircularDnaOptimizationProblem.resolve_constraints">
<code class="sig-name descname">resolve_constraints</code><span class="sig-paren">(</span><em class="sig-param">final_check=True</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dnachisel/CircularDnaOptimizationProblem.html#CircularDnaOptimizationProblem.resolve_constraints"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dnachisel.CircularDnaOptimizationProblem.resolve_constraints" title="Permalink to this definition">¶</a></dt>
<dd><p>Solve a particular constraint using local, targeted searches.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>constraint</strong></p>
<blockquote>
<div><p>The <code class="docutils literal notranslate"><span class="pre">Specification</span></code> object for which the sequence should be solved</p>
</div></blockquote>
<p><strong>final_check</strong></p>
<blockquote>
<div><p>If True, a final check of that all constraints pass will be run at
the end of the process, when constraints have been resolved one by
one, to check that the solving of one constraint didn’t undo the
solving of another.</p>
</div></blockquote>
<p><strong>cst_filter</strong></p>
<blockquote>
<div><p>An optional filter to only resolve a subset function (constraint =&gt; True/False)</p>
</div></blockquote>
</dd>
</dl>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="specification">
<h2><a class="toc-backref" href="#id4">Specification</a><a class="headerlink" href="#specification" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="dnachisel.Specification">
<em class="property">class </em><code class="sig-prename descclassname">dnachisel.</code><code class="sig-name descname">Specification</code><span class="sig-paren">(</span><em class="sig-param">evaluate=None</em>, <em class="sig-param">boost=1.0</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dnachisel/Specification.html#Specification"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dnachisel.Specification" title="Permalink to this definition">¶</a></dt>
<dd><p>General class to define specifications to optimize.</p>
<p>Note that all specifications have a <code class="docutils literal notranslate"><span class="pre">boost</span></code> attribute that is a
multiplicator that will be used when computing the global specification
score of a problem with <code class="docutils literal notranslate"><span class="pre">problem.all_objectives_score()</span></code>.</p>
<p>New types of specifications are defined by subclassing <code class="docutils literal notranslate"><span class="pre">Specification</span></code> and
providing a custom <code class="docutils literal notranslate"><span class="pre">evaluate</span></code> and <code class="docutils literal notranslate"><span class="pre">localized</span></code> methods.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>evaluate</strong></p>
<blockquote>
<div><p>function (sequence) =&gt; SpecEvaluation</p>
</div></blockquote>
<p><strong>boost</strong></p>
<blockquote>
<div><p>Relative importance of the Specification’s score in a multi-specification
problem.</p>
</div></blockquote>
</dd>
</dl>
<dl class="method">
<dt id="dnachisel.Specification.as_passive_objective">
<code class="sig-name descname">as_passive_objective</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dnachisel/Specification.html#Specification.as_passive_objective"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dnachisel.Specification.as_passive_objective" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a copy with optimize_passively set to true.</p>
<p>“Optimize passively” means that when the specification is used as an
objective, the solver will not do a specific pass to optimize this
specification, however this specification’s score will be taken into
account in the global score when optimizing other objectives, and
may therefore influence the final sequence.</p>
</dd></dl>

<dl class="method">
<dt id="dnachisel.Specification.copy_with_changes">
<code class="sig-name descname">copy_with_changes</code><span class="sig-paren">(</span><em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dnachisel/Specification.html#Specification.copy_with_changes"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dnachisel.Specification.copy_with_changes" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a copy of the Specification with modified properties.</p>
<p>For instance <code class="docutils literal notranslate"><span class="pre">new_spec</span> <span class="pre">=</span> <span class="pre">spec.copy_with_changes(boost=10)</span></code>.</p>
</dd></dl>

<dl class="method">
<dt id="dnachisel.Specification.from_biopython_feature">
<em class="property">static </em><code class="sig-name descname">from_biopython_feature</code><span class="sig-paren">(</span><em class="sig-param">feature</em>, <em class="sig-param">specifications_dict</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dnachisel/Specification.html#Specification.from_biopython_feature"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dnachisel.Specification.from_biopython_feature" title="Permalink to this definition">¶</a></dt>
<dd><p>Parse a Biopython feature create an annotation.</p>
<p>The specifications_dict enables to map specification names to the
actual implemented class.</p>
</dd></dl>

<dl class="method">
<dt id="dnachisel.Specification.initialized_on_problem">
<code class="sig-name descname">initialized_on_problem</code><span class="sig-paren">(</span><em class="sig-param">problem</em>, <em class="sig-param">role='constraint'</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dnachisel/Specification.html#Specification.initialized_on_problem"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dnachisel.Specification.initialized_on_problem" title="Permalink to this definition">¶</a></dt>
<dd><p>Complete specification initialization when the sequence gets known.</p>
<p>Some specifications like to know what their role is and on which
sequence they are employed before they complete some values.</p>
</dd></dl>

<dl class="method">
<dt id="dnachisel.Specification.label_parameters">
<code class="sig-name descname">label_parameters</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dnachisel/Specification.html#Specification.label_parameters"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dnachisel.Specification.label_parameters" title="Permalink to this definition">¶</a></dt>
<dd><p>In subclasses, returns a list of the creation parameters.</p>
<p>For instance [(‘pattern’, ‘ATT’), (‘occurences’, 2)]</p>
</dd></dl>

<dl class="method">
<dt id="dnachisel.Specification.localized">
<code class="sig-name descname">localized</code><span class="sig-paren">(</span><em class="sig-param">location</em>, <em class="sig-param">problem=None</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dnachisel/Specification.html#Specification.localized"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dnachisel.Specification.localized" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a modified version of the specification for the case where
sequence modifications are only performed inside the provided location.</p>
<p>For instance if an specification concerns local GC content, and we are
only making local mutations to destroy a restriction site, then we only
need to check the local GC content around the restriction site after
each mutation (and not compute it for the whole sequence), so
<code class="docutils literal notranslate"><span class="pre">EnforceGCContent.localized(location)</span></code> will return an specification
that only looks for GC content around the provided location.</p>
<p>If an specification concerns a DNA segment that is completely disjoint
from the provided location, this must return None.</p>
<p>Must return an object of class <code class="docutils literal notranslate"><span class="pre">Constraint</span></code>.</p>
</dd></dl>

<dl class="method">
<dt id="dnachisel.Specification.restrict_nucleotides">
<code class="sig-name descname">restrict_nucleotides</code><span class="sig-paren">(</span><em class="sig-param">sequence</em>, <em class="sig-param">location=None</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dnachisel/Specification.html#Specification.restrict_nucleotides"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dnachisel.Specification.restrict_nucleotides" title="Permalink to this definition">¶</a></dt>
<dd><p>Restrict the mutation space to speed up optimization.</p>
<p>This method only kicks in when this specification is used as a
constraint. By default it does nothing, but subclasses such as
EnforceTranslation, AvoidChanges, EnforceSequence, etc. have custom
methods.</p>
<p>In the code, this method is run during the initialize() step of
DNAOptimizationProblem, when the MutationSpace is created for each constraint</p>
</dd></dl>

<dl class="method">
<dt id="dnachisel.Specification.shifted">
<code class="sig-name descname">shifted</code><span class="sig-paren">(</span><em class="sig-param">shift</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dnachisel/Specification.html#Specification.shifted"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dnachisel.Specification.shifted" title="Permalink to this definition">¶</a></dt>
<dd><p>Shift the location of the specification.</p>
<p>Some specification classes may have a special method to do side effects
when shifting the location.</p>
<p>Location shifting is used in particular when solving circular DNA
optimization problems.</p>
</dd></dl>

<dl class="method">
<dt id="dnachisel.Specification.short_label">
<code class="sig-name descname">short_label</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dnachisel/Specification.html#Specification.short_label"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dnachisel.Specification.short_label" title="Permalink to this definition">¶</a></dt>
<dd><p>Shorter, less precise label to be used in tables, reports, etc.</p>
<p>This is meant for specifications such as EnforceGCContent(0.4, 0.6)
to be represented as ‘40-60% GC’ in reports tables etc..</p>
</dd></dl>

<dl class="method">
<dt id="dnachisel.Specification.to_biopython_feature">
<code class="sig-name descname">to_biopython_feature</code><span class="sig-paren">(</span><em class="sig-param">feature_type='misc_feature'</em>, <em class="sig-param">role='constraint'</em>, <em class="sig-param">colors_dict=None</em>, <em class="sig-param">**qualifiers</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dnachisel/Specification.html#Specification.to_biopython_feature"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dnachisel.Specification.to_biopython_feature" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a Biopython feature representing the specification.</p>
<p>The feature label is a string representation of the specification,
and its location indicates the specification’s scope.</p>
<p>This method is primarily meant to display specifications in Genbank.
They may result in “viable”, DnaChisel-compatible annotations that can
be imported back into DNA Chisel, but this is not the intended goal.</p>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="specevaluation">
<h2><a class="toc-backref" href="#id5">SpecEvaluation</a><a class="headerlink" href="#specevaluation" title="Permalink to this headline">¶</a></h2>
<span class="target" id="module-dnachisel.SpecEvaluation"></span><p>Implements the SpecEvaluation class</p>
<p>SpecEvaluation is a class describing the result of an the evaluation of a
<em>DnaOptimizationProblem</em> by a <em>Specification</em>. It contains a score, a message,
a list of <em>Locations</em> of sub-optimal regions.</p>
<p>Several evaluations can be grouped using classes
<em>ProblemConstraintsEvaluations</em> and <em>ProblemObjectivesEvaluations</em>, which
implement methods for printing or exporting as Genbank a set of evaluations.</p>
<dl class="class">
<dt id="dnachisel.SpecEvaluation.ProblemConstraintsEvaluations">
<em class="property">class </em><code class="sig-prename descclassname">dnachisel.SpecEvaluation.</code><code class="sig-name descname">ProblemConstraintsEvaluations</code><span class="sig-paren">(</span><em class="sig-param">evaluations</em>, <em class="sig-param">problem=None</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dnachisel/SpecEvaluation.html#ProblemConstraintsEvaluations"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dnachisel.SpecEvaluation.ProblemConstraintsEvaluations" title="Permalink to this definition">¶</a></dt>
<dd><p>Special multi-evaluation class for all constraints of a same problem.</p>
<p>See submethod <code class="docutils literal notranslate"><span class="pre">.from_problem</span></code></p>
<dl class="method">
<dt id="dnachisel.SpecEvaluation.ProblemConstraintsEvaluations.from_problem">
<em class="property">static </em><code class="sig-name descname">from_problem</code><span class="sig-paren">(</span><em class="sig-param">problem</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dnachisel/SpecEvaluation.html#ProblemConstraintsEvaluations.from_problem"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dnachisel.SpecEvaluation.ProblemConstraintsEvaluations.from_problem" title="Permalink to this definition">¶</a></dt>
<dd><p>Create an instance by evaluating all constraints in the problem.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">problem</span></code> is a DnaChisel DnaOptimizationProblem.</p>
</dd></dl>

<dl class="method">
<dt id="dnachisel.SpecEvaluation.ProblemConstraintsEvaluations.success_failure_color">
<code class="sig-name descname">success_failure_color</code><span class="sig-paren">(</span><em class="sig-param">evaluation</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dnachisel/SpecEvaluation.html#ProblemConstraintsEvaluations.success_failure_color"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dnachisel.SpecEvaluation.ProblemConstraintsEvaluations.success_failure_color" title="Permalink to this definition">¶</a></dt>
<dd><p>Return color #60f979 if evaluation.passes else #f96c60.</p>
</dd></dl>

<dl class="method">
<dt id="dnachisel.SpecEvaluation.ProblemConstraintsEvaluations.text_summary_message">
<code class="sig-name descname">text_summary_message</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dnachisel/SpecEvaluation.html#ProblemConstraintsEvaluations.text_summary_message"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dnachisel.SpecEvaluation.ProblemConstraintsEvaluations.text_summary_message" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a global SUCCESS or FAILURE message for all evaluations.</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="dnachisel.SpecEvaluation.ProblemObjectivesEvaluations">
<em class="property">class </em><code class="sig-prename descclassname">dnachisel.SpecEvaluation.</code><code class="sig-name descname">ProblemObjectivesEvaluations</code><span class="sig-paren">(</span><em class="sig-param">evaluations</em>, <em class="sig-param">problem=None</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dnachisel/SpecEvaluation.html#ProblemObjectivesEvaluations"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dnachisel.SpecEvaluation.ProblemObjectivesEvaluations" title="Permalink to this definition">¶</a></dt>
<dd><p>Special multi-evaluation class for all objectives of a same problem.</p>
<p>See submethod <code class="docutils literal notranslate"><span class="pre">.from_problem</span></code></p>
<dl class="method">
<dt id="dnachisel.SpecEvaluation.ProblemObjectivesEvaluations.from_problem">
<em class="property">static </em><code class="sig-name descname">from_problem</code><span class="sig-paren">(</span><em class="sig-param">problem</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dnachisel/SpecEvaluation.html#ProblemObjectivesEvaluations.from_problem"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dnachisel.SpecEvaluation.ProblemObjectivesEvaluations.from_problem" title="Permalink to this definition">¶</a></dt>
<dd><p>Create an instance by evaluating all objectives in the problem.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">problem</span></code> is a DnaChisel DnaOptimizationProblem.</p>
</dd></dl>

<dl class="method">
<dt id="dnachisel.SpecEvaluation.ProblemObjectivesEvaluations.success_failure_color">
<code class="sig-name descname">success_failure_color</code><span class="sig-paren">(</span><em class="sig-param">evaluation</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dnachisel/SpecEvaluation.html#ProblemObjectivesEvaluations.success_failure_color"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dnachisel.SpecEvaluation.ProblemObjectivesEvaluations.success_failure_color" title="Permalink to this definition">¶</a></dt>
<dd><p>Return color #cbf960 if evaluation is optimal else #f9a260.</p>
</dd></dl>

<dl class="method">
<dt id="dnachisel.SpecEvaluation.ProblemObjectivesEvaluations.text_summary_message">
<code class="sig-name descname">text_summary_message</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dnachisel/SpecEvaluation.html#ProblemObjectivesEvaluations.text_summary_message"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dnachisel.SpecEvaluation.ProblemObjectivesEvaluations.text_summary_message" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a TOTAL SCORE message.</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="dnachisel.SpecEvaluation.SpecEvaluation">
<em class="property">class </em><code class="sig-prename descclassname">dnachisel.SpecEvaluation.</code><code class="sig-name descname">SpecEvaluation</code><span class="sig-paren">(</span><em class="sig-param">specification</em>, <em class="sig-param">problem</em>, <em class="sig-param">score</em>, <em class="sig-param">locations=None</em>, <em class="sig-param">message=None</em>, <em class="sig-param">data=None</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dnachisel/SpecEvaluation.html#SpecEvaluation"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dnachisel.SpecEvaluation.SpecEvaluation" title="Permalink to this definition">¶</a></dt>
<dd><p>Store relevant infos about the evaluation of an objective on a problem.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>objective</strong></p>
<blockquote>
<div><p>The Objective that was evaluated.</p>
</div></blockquote>
<p><strong>problem</strong></p>
<blockquote>
<div><p>The problem that the objective was evaluated on.</p>
</div></blockquote>
<p><strong>score</strong></p>
<blockquote>
<div><p>The score associated to the evaluation.</p>
</div></blockquote>
<p><strong>locations</strong></p>
<blockquote>
<div><p>A list of couples (start, end) indicating the locations on which the
the optimization shoul be localized to improve the objective.</p>
</div></blockquote>
<p><strong>message</strong></p>
<blockquote>
<div><p>A message that will be returned by <code class="docutils literal notranslate"><span class="pre">str(evaluation)</span></code>. It will notably
be displayed by <code class="docutils literal notranslate"><span class="pre">problem.print_objectives_summaries</span></code>.</p>
</div></blockquote>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">evaluation_result</span> <span class="o">=</span> <span class="n">ConstraintEvaluation</span><span class="p">(</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">specification</span><span class="o">=</span><span class="n">specification</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">problem</span> <span class="o">=</span> <span class="n">problem</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">score</span><span class="o">=</span> <span class="n">evaluation_score</span><span class="p">,</span> <span class="c1"># float</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">locations</span><span class="o">=</span><span class="p">[</span><span class="n">Locations</span> <span class="nb">list</span><span class="o">...</span><span class="p">],</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">message</span> <span class="o">=</span> <span class="s2">&quot;Score: 42 (found 42 sites)&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="p">)</span>
</pre></div>
</div>
<dl class="method">
<dt id="dnachisel.SpecEvaluation.SpecEvaluation.default_message">
<em class="property">property </em><code class="sig-name descname">default_message</code><a class="headerlink" href="#dnachisel.SpecEvaluation.SpecEvaluation.default_message" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the default message for console/reports.</p>
</dd></dl>

<dl class="method">
<dt id="dnachisel.SpecEvaluation.SpecEvaluation.locations_to_biopython_features">
<code class="sig-name descname">locations_to_biopython_features</code><span class="sig-paren">(</span><em class="sig-param">feature_type='misc_feature'</em>, <em class="sig-param">color='red'</em>, <em class="sig-param">label_prefix=''</em>, <em class="sig-param">merge_overlapping=False</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dnachisel/SpecEvaluation.html#SpecEvaluation.locations_to_biopython_features"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dnachisel.SpecEvaluation.SpecEvaluation.locations_to_biopython_features" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a list of locations (of breach/suboptimality) as annotations.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>feature_type</strong></p>
<blockquote>
<div><p>Genbank type of the annotations</p>
</div></blockquote>
<p><strong>color</strong></p>
<blockquote>
<div><p>Color property attached to the annotations</p>
</div></blockquote>
<p><strong>label_prefix</strong></p>
<blockquote>
<div><p>The locations will be labelled of the form
“prefix NameOfSpecification()”</p>
</div></blockquote>
<p><strong>merge_overlapping</strong></p>
<blockquote>
<div><p>If true, then overlapping locations (0-5, 2-9) will be merged into a
single one (0-9).</p>
</div></blockquote>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="dnachisel.SpecEvaluation.SpecEvaluation.to_text">
<code class="sig-name descname">to_text</code><span class="sig-paren">(</span><em class="sig-param">role='constraint'</em>, <em class="sig-param">wrapped=True</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dnachisel/SpecEvaluation.html#SpecEvaluation.to_text"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dnachisel.SpecEvaluation.SpecEvaluation.to_text" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a string representation of the evaluation.</p>
<p>Example output for a constraint:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>&gt;&gt;&gt; FAIL ┍ AvoidNonUniqueSegments[10-1000](min_length:9)
&gt;&gt;&gt;      │ Score:        -2. Locations: [232-241, 233-242]
</pre></div>
</div>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>role</strong></p>
<blockquote>
<div><p>either ‘objective’ or ‘constraint’</p>
</div></blockquote>
<p><strong>wrapped</strong></p>
<blockquote>
<div><p>Whether to wrap-and-indent the result</p>
</div></blockquote>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>message</p>
<blockquote>
<div><p>A long string detailing the evaluations’s status and regions of
breach or suboptimality.</p>
</div></blockquote>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="dnachisel.SpecEvaluation.SpecEvaluations">
<em class="property">class </em><code class="sig-prename descclassname">dnachisel.SpecEvaluation.</code><code class="sig-name descname">SpecEvaluations</code><span class="sig-paren">(</span><em class="sig-param">evaluations</em>, <em class="sig-param">problem=None</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dnachisel/SpecEvaluation.html#SpecEvaluations"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dnachisel.SpecEvaluation.SpecEvaluations" title="Permalink to this definition">¶</a></dt>
<dd><p>Base class for handling lists of SpecEvaluations.</p>
<p>See ProblemObjectivesEvaluations and ProblemConstraintsEvaluations for
the useful subclasses.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>evaluations</strong></p>
<blockquote>
<div><p>list of SpecEvaluations</p>
</div></blockquote>
<p><strong>problem</strong></p>
<blockquote>
<div><p>(optional) problem on which the evaluations were carried.</p>
</div></blockquote>
</dd>
</dl>
<dl class="method">
<dt id="dnachisel.SpecEvaluation.SpecEvaluations.all_evaluations_pass">
<code class="sig-name descname">all_evaluations_pass</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dnachisel/SpecEvaluation.html#SpecEvaluations.all_evaluations_pass"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dnachisel.SpecEvaluation.SpecEvaluations.all_evaluations_pass" title="Permalink to this definition">¶</a></dt>
<dd><p>Return whether all evaluations pass.</p>
</dd></dl>

<dl class="method">
<dt id="dnachisel.SpecEvaluation.SpecEvaluations.evaluations_with_locations">
<code class="sig-name descname">evaluations_with_locations</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dnachisel/SpecEvaluation.html#SpecEvaluations.evaluations_with_locations"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dnachisel.SpecEvaluation.SpecEvaluations.evaluations_with_locations" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the list of all evaluations whose location is not None.</p>
</dd></dl>

<dl class="method">
<dt id="dnachisel.SpecEvaluation.SpecEvaluations.filter">
<code class="sig-name descname">filter</code><span class="sig-paren">(</span><em class="sig-param">eval_filter</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dnachisel/SpecEvaluation.html#SpecEvaluations.filter"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dnachisel.SpecEvaluation.SpecEvaluations.filter" title="Permalink to this definition">¶</a></dt>
<dd><p>Create a new instance with a subset of the evaluations.</p>
<p><code class="docutils literal notranslate"><span class="pre">eval_filter</span></code> is either a function <code class="docutils literal notranslate"><span class="pre">(evaluation)</span> <span class="pre">=&gt;</span> <span class="pre">True/False</span></code>
or one of “passing”, “failing”, “optimal”, “suboptimal”, to obtain the
corresponding constraints.</p>
</dd></dl>

<dl class="method">
<dt id="dnachisel.SpecEvaluation.SpecEvaluations.locations_as_features">
<code class="sig-name descname">locations_as_features</code><span class="sig-paren">(</span><em class="sig-param">features_type='misc_feature'</em>, <em class="sig-param">with_specifications=True</em>, <em class="sig-param">label_prefix='From'</em>, <em class="sig-param">colors='cycle'</em>, <em class="sig-param">merge_overlapping=False</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dnachisel/SpecEvaluation.html#SpecEvaluations.locations_as_features"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dnachisel.SpecEvaluation.SpecEvaluations.locations_as_features" title="Permalink to this definition">¶</a></dt>
<dd><p>Return all locations from all evaluations as biopython features.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>features_type</strong></p>
<blockquote>
<div><p>The Genbank feature type</p>
</div></blockquote>
<p><strong>with_specifications</strong></p>
<blockquote>
<div><p>If True, features are added to the list to indicate the scope of the
different specifications. If false, only the specification breaches
are returned.</p>
</div></blockquote>
<p><strong>label_prefix</strong></p>
<blockquote>
<div><p>Each breach may be labeled “prefix NameOfSpec(props)”, for instance,
“From AvoidPattern(100-200)”, to indicate where the breach belongs.</p>
</div></blockquote>
<p><strong>colors</strong></p>
<blockquote>
<div><p>Either a list of colors (one for each specification), e.g.
[‘red’, ‘#7aab71’, …] or “cycle” for cycling through predefined
colors. The colors are applied to all breaches.</p>
</div></blockquote>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="dnachisel.SpecEvaluation.SpecEvaluations.scores_sum">
<code class="sig-name descname">scores_sum</code><span class="sig-paren">(</span><em class="sig-param">as_text=False</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dnachisel/SpecEvaluation.html#SpecEvaluations.scores_sum"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dnachisel.SpecEvaluation.SpecEvaluations.scores_sum" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the sum of all evaluations scores.</p>
<p>Scores are multiplied by their respective boost factor.</p>
</dd></dl>

<dl class="method">
<dt id="dnachisel.SpecEvaluation.SpecEvaluations.success_and_failures_as_features">
<code class="sig-name descname">success_and_failures_as_features</code><span class="sig-paren">(</span><em class="sig-param">feature_type='misc_feature'</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dnachisel/SpecEvaluation.html#SpecEvaluations.success_and_failures_as_features"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dnachisel.SpecEvaluation.SpecEvaluations.success_and_failures_as_features" title="Permalink to this definition">¶</a></dt>
<dd><p>Return all evaluations as Biopython features.</p>
<p>With color property depending on whether the evaluation is passing,
failing, optimal, or suboptimal, the color being determined by
<code class="docutils literal notranslate"><span class="pre">self.success_failure_color</span></code>.</p>
</dd></dl>

<dl class="method">
<dt id="dnachisel.SpecEvaluation.SpecEvaluations.to_text">
<code class="sig-name descname">to_text</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dnachisel/SpecEvaluation.html#SpecEvaluations.to_text"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dnachisel.SpecEvaluation.SpecEvaluations.to_text" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a long representation of the evaluations list.</p>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-dnachisel.Location">
<span id="location"></span><h2><a class="toc-backref" href="#id6">Location</a><a class="headerlink" href="#module-dnachisel.Location" title="Permalink to this headline">¶</a></h2>
<p>Implements the Location class.</p>
<p>The class has useful methods for finding overlaps between locations, extract
a subsequence from a sequence, etc.</p>
<dl class="class">
<dt id="dnachisel.Location.Location">
<em class="property">class </em><code class="sig-prename descclassname">dnachisel.Location.</code><code class="sig-name descname">Location</code><span class="sig-paren">(</span><em class="sig-param">start</em>, <em class="sig-param">end</em>, <em class="sig-param">strand=None</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dnachisel/Location.html#Location"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dnachisel.Location.Location" title="Permalink to this definition">¶</a></dt>
<dd><p>Represent a segment of a sequence, with a start, end, and strand.</p>
<p>Warning: we use Python’s splicing notation, so Location(5, 10) represents
sequence[5, 6, 7, 8, 9] which corresponds to nucleotides number
6, 7, 8, 9, 10.</p>
<p>The data structure is similar to a Biopython’s FeatureLocation, but with
different methods for different purposes.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>start</strong></p>
<blockquote>
<div><p>Lowest position index of the segment</p>
</div></blockquote>
<p><strong>end</strong></p>
<blockquote>
<div><p>Highest position index of the segment</p>
</div></blockquote>
<p><strong>strand</strong></p>
<blockquote>
<div><p>Either 1 or -1 for sense or anti-sense orientation.</p>
</div></blockquote>
</dd>
</dl>
<dl class="method">
<dt id="dnachisel.Location.Location.extended">
<code class="sig-name descname">extended</code><span class="sig-paren">(</span><em class="sig-param">extension_length</em>, <em class="sig-param">lower_limit=0</em>, <em class="sig-param">upper_limit=None</em>, <em class="sig-param">left=True</em>, <em class="sig-param">right=True</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dnachisel/Location.html#Location.extended"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dnachisel.Location.Location.extended" title="Permalink to this definition">¶</a></dt>
<dd><p>Extend the location of a few basepairs on each side.</p>
</dd></dl>

<dl class="method">
<dt id="dnachisel.Location.Location.extract_sequence">
<code class="sig-name descname">extract_sequence</code><span class="sig-paren">(</span><em class="sig-param">sequence</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dnachisel/Location.html#Location.extract_sequence"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dnachisel.Location.Location.extract_sequence" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the subsequence read at the given location.</p>
</dd></dl>

<dl class="method">
<dt id="dnachisel.Location.Location.from_biopython_location">
<em class="property">static </em><code class="sig-name descname">from_biopython_location</code><span class="sig-paren">(</span><em class="sig-param">location</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dnachisel/Location.html#Location.from_biopython_location"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dnachisel.Location.Location.from_biopython_location" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a DnaChisel Location from a Biopython location.</p>
</dd></dl>

<dl class="method">
<dt id="dnachisel.Location.Location.merge_overlapping_locations">
<em class="property">static </em><code class="sig-name descname">merge_overlapping_locations</code><span class="sig-paren">(</span><em class="sig-param">locations</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dnachisel/Location.html#Location.merge_overlapping_locations"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dnachisel.Location.Location.merge_overlapping_locations" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a list of locations obtained by mergin all overlapping.</p>
</dd></dl>

<dl class="method">
<dt id="dnachisel.Location.Location.overlap_region">
<code class="sig-name descname">overlap_region</code><span class="sig-paren">(</span><em class="sig-param">other_location</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dnachisel/Location.html#Location.overlap_region"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dnachisel.Location.Location.overlap_region" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the overlap span between two locations (None if None).</p>
</dd></dl>

<dl class="method">
<dt id="dnachisel.Location.Location.to_biopython_feature">
<code class="sig-name descname">to_biopython_feature</code><span class="sig-paren">(</span><em class="sig-param">feature_type='misc_feature'</em>, <em class="sig-param">**qualifiers</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dnachisel/Location.html#Location.to_biopython_feature"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dnachisel.Location.Location.to_biopython_feature" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a Biopython SeqFeature with same location and custom
qualifiers.</p>
</dd></dl>

<dl class="method">
<dt id="dnachisel.Location.Location.to_biopython_location">
<code class="sig-name descname">to_biopython_location</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dnachisel/Location.html#Location.to_biopython_location"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dnachisel.Location.Location.to_biopython_location" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a Biopython FeatureLocation equivalent to the location.</p>
</dd></dl>

<dl class="method">
<dt id="dnachisel.Location.Location.to_tuple">
<code class="sig-name descname">to_tuple</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dnachisel/Location.html#Location.to_tuple"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dnachisel.Location.Location.to_tuple" title="Permalink to this definition">¶</a></dt>
<dd><p>Return (start, end, strand).</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="dnachisel.Location.MultiLocation">
<em class="property">class </em><code class="sig-prename descclassname">dnachisel.Location.</code><code class="sig-name descname">MultiLocation</code><span class="sig-paren">(</span><em class="sig-param">locations</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dnachisel/Location.html#MultiLocation"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dnachisel.Location.MultiLocation" title="Permalink to this definition">¶</a></dt>
<dd><p>TODO: Implement when necessary</p>
</dd></dl>

</div>
<div class="section" id="module-dnachisel.MutationSpace">
<span id="mutationspace"></span><h2><a class="toc-backref" href="#id7">MutationSpace</a><a class="headerlink" href="#module-dnachisel.MutationSpace" title="Permalink to this headline">¶</a></h2>
<p>Define MutationSpace</p>
<dl class="class">
<dt id="dnachisel.MutationSpace.MutationChoice">
<em class="property">class </em><code class="sig-prename descclassname">dnachisel.MutationSpace.</code><code class="sig-name descname">MutationChoice</code><span class="sig-paren">(</span><em class="sig-param">segment</em>, <em class="sig-param">variants</em>, <em class="sig-param">is_any_nucleotide=False</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dnachisel/MutationSpace.html#MutationChoice"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dnachisel.MutationSpace.MutationChoice" title="Permalink to this definition">¶</a></dt>
<dd><p>Represent a segment of a sequence with several possible variants.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>segment</strong></p>
<blockquote>
<div><p>A pair (start, end) indicating the range of nucleotides concerned. We
are applying Python range, so</p>
</div></blockquote>
<p><strong>variants</strong></p>
<blockquote>
<div><p>A set of sequence variants, at the given position</p>
</div></blockquote>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">choice</span> <span class="o">=</span> <span class="n">MutationChoice</span><span class="p">((</span><span class="mi">70</span><span class="p">,</span> <span class="mi">73</span><span class="p">),</span> <span class="p">{})</span>
</pre></div>
</div>
<dl class="method">
<dt id="dnachisel.MutationSpace.MutationChoice.extract_varying_region">
<code class="sig-name descname">extract_varying_region</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dnachisel/MutationSpace.html#MutationChoice.extract_varying_region"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dnachisel.MutationSpace.MutationChoice.extract_varying_region" title="Permalink to this definition">¶</a></dt>
<dd><p>Return MutationChoices for the central varying region and 2 flanks.</p>
<p>For instance:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">choice</span> <span class="o">=</span> <span class="n">MutationChoice</span><span class="p">((</span><span class="mi">5</span><span class="p">,</span> <span class="mi">12</span><span class="p">),</span> <span class="p">[</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="s1">&#39;ATGCGTG&#39;</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="s1">&#39;AAAAATG&#39;</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="s1">&#39;AAATGTG&#39;</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="s1">&#39;ATGAATG&#39;</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">choice</span><span class="o">.</span><span class="n">extract_varying_region</span><span class="p">()</span>
</pre></div>
</div>
<p>Result :</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="p">[</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">MutChoice</span><span class="p">(</span><span class="mi">5</span><span class="o">-</span><span class="mi">6</span> <span class="n">A</span><span class="p">),</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">MutChoice</span><span class="p">(</span><span class="mi">6</span><span class="o">-</span><span class="mi">10</span> <span class="n">TGCG</span><span class="o">-</span><span class="n">AATG</span><span class="o">-</span><span class="n">TGAA</span><span class="o">-</span><span class="n">AAAA</span><span class="p">),</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">MutChoice</span><span class="p">(</span><span class="mi">10</span><span class="o">-</span><span class="mi">12</span> <span class="n">TG</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="p">]</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="dnachisel.MutationSpace.MutationChoice.merge_with">
<code class="sig-name descname">merge_with</code><span class="sig-paren">(</span><em class="sig-param">others</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dnachisel/MutationSpace.html#MutationChoice.merge_with"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dnachisel.MutationSpace.MutationChoice.merge_with" title="Permalink to this definition">¶</a></dt>
<dd><p>Merge this mutation choice with others to form a single choice</p>
</dd></dl>

<dl class="method">
<dt id="dnachisel.MutationSpace.MutationChoice.random_variant">
<code class="sig-name descname">random_variant</code><span class="sig-paren">(</span><em class="sig-param">sequence</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dnachisel/MutationSpace.html#MutationChoice.random_variant"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dnachisel.MutationSpace.MutationChoice.random_variant" title="Permalink to this definition">¶</a></dt>
<dd><p>Return one of the variants, randomly.</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="dnachisel.MutationSpace.MutationSpace">
<em class="property">class </em><code class="sig-prename descclassname">dnachisel.MutationSpace.</code><code class="sig-name descname">MutationSpace</code><span class="sig-paren">(</span><em class="sig-param">choices_index</em>, <em class="sig-param">left_padding=0</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dnachisel/MutationSpace.html#MutationSpace"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dnachisel.MutationSpace.MutationSpace" title="Permalink to this definition">¶</a></dt>
<dd><p>Class for mutation space (set of sequence segments and their variants).</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>choices_index</strong></p>
<blockquote>
<div><p>A list L such that L[i] gives the MutationChoice governing the mutations
allowed at position i (ansd possibly around i)</p>
</div></blockquote>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="c1"># BEWARE: below, similar mutation choices are actually the SAME OBJECT</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">space</span> <span class="o">=</span> <span class="n">MutationSpace</span><span class="p">([</span>
<span class="go">        MutationChoice((0, 2), {&#39;AT&#39;, &#39;TG&#39;}),</span>
<span class="go">        MutationChoice((0, 2), {&#39;AT&#39;, &#39;TG&#39;}),</span>
<span class="go">        MutationChoice((2, 5), {&#39;TTC&#39;, &#39;TTA&#39;, &#39;TTT&#39;}), # same</span>
<span class="go">        MutationChoice((2, 5), {&#39;TTC&#39;, &#39;TTA&#39;, &#39;TTT&#39;}), #</span>
<span class="go">        MutationChoice((2, 5), {&#39;TTC&#39;, &#39;TTA&#39;, &#39;TTT&#39;}),</span>
<span class="go">    ])</span>
</pre></div>
</div>
<dl class="method">
<dt id="dnachisel.MutationSpace.MutationSpace.all_variants">
<code class="sig-name descname">all_variants</code><span class="sig-paren">(</span><em class="sig-param">sequence</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dnachisel/MutationSpace.html#MutationSpace.all_variants"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dnachisel.MutationSpace.MutationSpace.all_variants" title="Permalink to this definition">¶</a></dt>
<dd><p>Iterate through all sequence variants in this mutation space.</p>
</dd></dl>

<dl class="method">
<dt id="dnachisel.MutationSpace.MutationSpace.apply_random_mutations">
<code class="sig-name descname">apply_random_mutations</code><span class="sig-paren">(</span><em class="sig-param">n_mutations</em>, <em class="sig-param">sequence</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dnachisel/MutationSpace.html#MutationSpace.apply_random_mutations"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dnachisel.MutationSpace.MutationSpace.apply_random_mutations" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a sequence with n random mutations applied.</p>
</dd></dl>

<dl class="method">
<dt id="dnachisel.MutationSpace.MutationSpace.choices_span">
<em class="property">property </em><code class="sig-name descname">choices_span</code><a class="headerlink" href="#dnachisel.MutationSpace.MutationSpace.choices_span" title="Permalink to this definition">¶</a></dt>
<dd><p>Return (start, end), segment where mutiple choices are possible</p>
</dd></dl>

<dl class="method">
<dt id="dnachisel.MutationSpace.MutationSpace.constrain_sequence">
<code class="sig-name descname">constrain_sequence</code><span class="sig-paren">(</span><em class="sig-param">sequence</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dnachisel/MutationSpace.html#MutationSpace.constrain_sequence"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dnachisel.MutationSpace.MutationSpace.constrain_sequence" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a version of the sequence compatible with the mutation space.</p>
<p>All nucleotides of the sequence that are incompatible with the
mutation space are replaced by nucleotides compatible with the space.</p>
</dd></dl>

<dl class="method">
<dt id="dnachisel.MutationSpace.MutationSpace.from_optimization_problem">
<em class="property">static </em><code class="sig-name descname">from_optimization_problem</code><span class="sig-paren">(</span><em class="sig-param">problem</em>, <em class="sig-param">new_constraints=None</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dnachisel/MutationSpace.html#MutationSpace.from_optimization_problem"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dnachisel.MutationSpace.MutationSpace.from_optimization_problem" title="Permalink to this definition">¶</a></dt>
<dd><p>Create a mutation space from a DNA optimization problem.</p>
<p>This can be used either to initialize mutation spaces for new problems,
or to</p>
</dd></dl>

<dl class="method">
<dt id="dnachisel.MutationSpace.MutationSpace.localized">
<code class="sig-name descname">localized</code><span class="sig-paren">(</span><em class="sig-param">location</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dnachisel/MutationSpace.html#MutationSpace.localized"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dnachisel.MutationSpace.MutationSpace.localized" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a new version with only mutations overlaping the location.</p>
</dd></dl>

<dl class="method">
<dt id="dnachisel.MutationSpace.MutationSpace.pick_random_mutations">
<code class="sig-name descname">pick_random_mutations</code><span class="sig-paren">(</span><em class="sig-param">n_mutations</em>, <em class="sig-param">sequence</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dnachisel/MutationSpace.html#MutationSpace.pick_random_mutations"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dnachisel.MutationSpace.MutationSpace.pick_random_mutations" title="Permalink to this definition">¶</a></dt>
<dd><p>Draw N random mutations</p>
</dd></dl>

<dl class="method">
<dt id="dnachisel.MutationSpace.MutationSpace.space_size">
<em class="property">property </em><code class="sig-name descname">space_size</code><a class="headerlink" href="#dnachisel.MutationSpace.MutationSpace.space_size" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the number of possible mutations</p>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="sequencepattern">
<h2><a class="toc-backref" href="#id8">SequencePattern</a><a class="headerlink" href="#sequencepattern" title="Permalink to this headline">¶</a></h2>
<p>Specifications <code class="docutils literal notranslate"><span class="pre">AvoidPattern</span></code>, <code class="docutils literal notranslate"><span class="pre">EnforcePatternOccurence</span></code> accept a Pattern
as argument. The pattern can be provided either as a string or as a class:</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 25%" />
<col style="width: 46%" />
<col style="width: 28%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Name</p></th>
<th class="head"><p>Pattern Class</p></th>
<th class="head"><p>As string</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>Restriction site</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">EnzymeSitePattern('BsmBI')</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">'BsmBI_site'</span></code></p></td>
</tr>
<tr class="row-odd"><td><p>Homopolymer</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">HomopolymerPattern(&quot;A&quot;,</span> <span class="pre">6)</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">'6xA'</span></code></p></td>
</tr>
<tr class="row-even"><td><p>Direct repeats</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">RepeatedKmerPattern(3,</span> <span class="pre">2)</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">'3x2mer'</span></code></p></td>
</tr>
<tr class="row-odd"><td><p>DNA notation</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">DnaNotationPattern('ANNKA')</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">'ANNKA'</span></code></p></td>
</tr>
<tr class="row-even"><td><p>Regular Expression</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">SequencePattern('A[CG]*A')</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">'A[CG]*A'</span></code></p></td>
</tr>
</tbody>
</table>
<span class="target" id="module-dnachisel.SequencePattern"></span><p>Implements the SequencePattern, DnaNotationPattern classes.</p>
<p>These classes are responsible for looking for a pattern in a sequence
(including overlapping patterns !), separating patterns with fixed size
and patterns with maximal size (from problem localization purposes).</p>
<p>The module also implements functions to specify common DNA patterns:
homopolymers, repeats, enzymatic restriction sites.</p>
<dl class="class">
<dt id="dnachisel.SequencePattern.DnaNotationPattern">
<em class="property">class </em><code class="sig-prename descclassname">dnachisel.SequencePattern.</code><code class="sig-name descname">DnaNotationPattern</code><span class="sig-paren">(</span><em class="sig-param">sequence</em>, <em class="sig-param">name=None</em>, <em class="sig-param">in_both_strands='auto'</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dnachisel/SequencePattern.html#DnaNotationPattern"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dnachisel.SequencePattern.DnaNotationPattern" title="Permalink to this definition">¶</a></dt>
<dd><p>Class for patterns in plain DNA notation: ATTGCCA, GCNNKTA, etc.</p>
<p>If the sequence is not palyndromic, the pattern will be looked for in
both strands of sequences.</p>
<dl class="method">
<dt id="dnachisel.SequencePattern.DnaNotationPattern.all_variants">
<code class="sig-name descname">all_variants</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dnachisel/SequencePattern.html#DnaNotationPattern.all_variants"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dnachisel.SequencePattern.DnaNotationPattern.all_variants" title="Permalink to this definition">¶</a></dt>
<dd><p>Return all ATGC sequence variants of a sequence</p>
</dd></dl>

<dl class="method">
<dt id="dnachisel.SequencePattern.DnaNotationPattern.dna_sequence_to_regexpr">
<em class="property">static </em><code class="sig-name descname">dna_sequence_to_regexpr</code><span class="sig-paren">(</span><em class="sig-param">sequence</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dnachisel/SequencePattern.html#DnaNotationPattern.dna_sequence_to_regexpr"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dnachisel.SequencePattern.DnaNotationPattern.dna_sequence_to_regexpr" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a regular expression to find the pattern in a sequence.</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="dnachisel.SequencePattern.EnzymeSitePattern">
<em class="property">class </em><code class="sig-prename descclassname">dnachisel.SequencePattern.</code><code class="sig-name descname">EnzymeSitePattern</code><span class="sig-paren">(</span><em class="sig-param">enzyme_name</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dnachisel/SequencePattern.html#EnzymeSitePattern"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dnachisel.SequencePattern.EnzymeSitePattern" title="Permalink to this definition">¶</a></dt>
<dd><p>Class to represent Enzyme site patterns</p>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">enzyme_pattern</span> <span class="o">=</span> <span class="n">EnzymeSitePattern</span><span class="p">(</span><span class="s2">&quot;BsaI&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">constraint</span> <span class="o">=</span> <span class="n">AvoidPattern</span><span class="p">(</span><span class="n">enzyme_pattern</span><span class="p">)</span>
</pre></div>
</div>
<dl class="method">
<dt id="dnachisel.SequencePattern.EnzymeSitePattern.from_string">
<em class="property">static </em><code class="sig-name descname">from_string</code><span class="sig-paren">(</span><em class="sig-param">string</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dnachisel/SequencePattern.html#EnzymeSitePattern.from_string"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dnachisel.SequencePattern.EnzymeSitePattern.from_string" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert BsmBI_site to EnzymeSitePattern(BsmBI)</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="dnachisel.SequencePattern.HomopolymerPattern">
<em class="property">class </em><code class="sig-prename descclassname">dnachisel.SequencePattern.</code><code class="sig-name descname">HomopolymerPattern</code><span class="sig-paren">(</span><em class="sig-param">nucleotide</em>, <em class="sig-param">number</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dnachisel/SequencePattern.html#HomopolymerPattern"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dnachisel.SequencePattern.HomopolymerPattern" title="Permalink to this definition">¶</a></dt>
<dd><p>Homopolymer of the form AAAAAAA, TTTTT, etc.</p>
<p>Shorthand string version: “7xA”, “9xC”, etc.</p>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">pattern</span> <span class="o">=</span> <span class="n">HomopolymerPattern</span><span class="p">(</span><span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="mi">6</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">constraint</span> <span class="o">=</span> <span class="n">AvoidPattern</span><span class="p">(</span><span class="n">pattern</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="class">
<dt id="dnachisel.SequencePattern.RepeatedKmerPattern">
<em class="property">class </em><code class="sig-prename descclassname">dnachisel.SequencePattern.</code><code class="sig-name descname">RepeatedKmerPattern</code><span class="sig-paren">(</span><em class="sig-param">n_repeats</em>, <em class="sig-param">kmer_size</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dnachisel/SequencePattern.html#RepeatedKmerPattern"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dnachisel.SequencePattern.RepeatedKmerPattern" title="Permalink to this definition">¶</a></dt>
<dd><p>Direct repeats like ATT-ATT, ATGC-ATGC-ATGC, etc.</p>
<p>Shorthand string version: “3x4mer”, “5x2mer”, etc.</p>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">RepeatedKmerPattern</span><span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="mi">2</span><span class="p">)</span> <span class="c1"># dimers repeated 3 times</span>
</pre></div>
</div>
</dd></dl>

<dl class="class">
<dt id="dnachisel.SequencePattern.SequencePattern">
<em class="property">class </em><code class="sig-prename descclassname">dnachisel.SequencePattern.</code><code class="sig-name descname">SequencePattern</code><span class="sig-paren">(</span><em class="sig-param">expression</em>, <em class="sig-param">size=None</em>, <em class="sig-param">name=None</em>, <em class="sig-param">in_both_strands=True</em>, <em class="sig-param">lookahead='loop'</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dnachisel/SequencePattern.html#SequencePattern"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dnachisel.SequencePattern.SequencePattern" title="Permalink to this definition">¶</a></dt>
<dd><p>Pattern/ that will be looked for in a DNA sequence.</p>
<p>Use this class for matching regular expression patterns, and
DnaNotationPattern for matching explicit sequences or sequences using Ns
etc.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>expression</strong></p>
<blockquote>
<div><p>Any string or regular expression for matching ATGC nucleotides.
Note that multi-nucleotides symbols such as “N” (for A-T-G-C), or “K”
are not supported by this class, see DnaNotationPattern instead.</p>
</div></blockquote>
<p><strong>size</strong></p>
<blockquote>
<div><p>Size of the pattern, in number of characters (if none provided, the size
of the <code class="docutils literal notranslate"><span class="pre">pattern</span></code> string is used).
The <code class="docutils literal notranslate"><span class="pre">size</span></code> is used to determine the size of windows when performing
local optimization and constraint solving.
It can be important to provide the size when the
<code class="docutils literal notranslate"><span class="pre">pattern</span></code> string provided represents a complex regular expression whose
maximal matching size cannot be easily evaluated.</p>
</div></blockquote>
<p><strong>name</strong></p>
<blockquote>
<div><p>Name of the pattern (will be displayed e.g. when the pattern is printed)</p>
</div></blockquote>
<p><strong>in_both_strands</strong></p>
<blockquote>
<div><p>Set to True (default) if the pattern should also be looked for on the
reverse-complement of sequences.</p>
</div></blockquote>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">expression</span> <span class="o">=</span> <span class="s2">&quot;A[ATGC]{3,}&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pattern</span> <span class="o">=</span> <span class="n">SequencePattern</span><span class="p">(</span><span class="n">expression</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">constraint</span> <span class="o">=</span> <span class="n">AvoidPattern</span><span class="p">(</span><span class="n">pattern</span><span class="p">)</span>
</pre></div>
</div>
<dl class="method">
<dt id="dnachisel.SequencePattern.SequencePattern.find_matches">
<code class="sig-name descname">find_matches</code><span class="sig-paren">(</span><em class="sig-param">sequence</em>, <em class="sig-param">location=None</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dnachisel/SequencePattern.html#SequencePattern.find_matches"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dnachisel.SequencePattern.SequencePattern.find_matches" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the locations where the sequence matches the expression.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>sequence</strong></p>
<blockquote>
<div><p>A string of “ATGC…”</p>
</div></blockquote>
<p><strong>location</strong></p>
<blockquote>
<div><p>Location indicating a segment to which to restrict
the search. Only patterns entirely included in the segment will be
returned</p>
</div></blockquote>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>matches</p>
<blockquote>
<div><p>A list of the locations of matches, of the form
<code class="docutils literal notranslate"><span class="pre">[(start1,</span> <span class="pre">end1),</span> <span class="pre">(start2,</span> <span class="pre">end2),...]</span></code>.</p>
</div></blockquote>
</dd>
</dl>
</dd></dl>

</dd></dl>

</div>
</div>


          </div>
          <div class="page-nav">
            <div class="inner"><ul class="page-nav">
  <li class="prev">
    <a href="ref.html"
       title="previous chapter">← DNA Chisel API reference</a>
  </li>
  <li class="next">
    <a href="reports.html"
       title="next chapter">Reports →</a>
  </li>
</ul><div class="footer" role="contentinfo">
      &#169; Copyright 2019, Edinburgh Genome Foundry.
    <br>
    Created using <a href="http://sphinx-doc.org/">Sphinx</a> 2.2.0 with <a href="https://github.com/schettino72/sphinx_press_theme">Press Theme</a>.
</div>
            </div>
          </div>
      </page>
  </div></body>
</html>