

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>dnachisel.SequencePattern &mdash; DnaChisel 0.1.0 documentation</title>
  

  
  
  
  

  

  
  
    

  

  
  
    <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
  

  
    <link rel="stylesheet" href="../../_static/css/main.css" type="text/css" />
  
    <link rel="stylesheet" href="https://cdn.rawgit.com/knsv/mermaid/7.0.0/dist/mermaid.css" type="text/css" />
  

  
        <link rel="index" title="Index"
              href="../../genindex.html"/>
        <link rel="search" title="Search" href="../../search.html"/>
    <link rel="top" title="DnaChisel 0.1.0 documentation" href="../../index.html"/>
        <link rel="up" title="Module code" href="../index.html"/> 

  
  <script src="../../_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav" role="document">

   
  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="../../index.html" class="icon icon-home"> DnaChisel
          

          
            
            <img src="../../_static/html_logo.png" class="logo" />
          
          </a>

          
            
            
              <div class="version">
                0.1.0
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../../index.html">DnaChisel</a></li>
</ul>
<p class="caption"><span class="caption-text">Reference</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../ref/ref.html">DnaChisel Reference manual</a></li>
</ul>
<p class="caption"><span class="caption-text">Examples</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../examples/plasmid_optimization.html">Plasmid optimization with DnaChisel</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../examples/non_unique_kmers_minimization.html">Minimization of the number of repeated kmers</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../examples/pattern_instertion.html">Pattern insertion in a DNA sequence</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">DnaChisel</a>
        
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../index.html">Docs</a> &raquo;</li>
        
          <li><a href="../index.html">Module code</a> &raquo;</li>
        
      <li>dnachisel.SequencePattern</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <h1>Source code for dnachisel.SequencePattern</h1><div class="highlight"><pre>
<span></span><span class="sd">&quot;&quot;&quot;Implements the SequencePattern, DnaNotationPattern classes.</span>

<span class="sd">These classes are responsible for looking for a pattern in a sequence</span>
<span class="sd">(including overlapping patterns !), separating patterns with fixed size</span>
<span class="sd">and patterns with maximal size (from problem localization purposes).</span>

<span class="sd">The module also implements functions to specify common DNA patterns:</span>
<span class="sd">homopolymers, repeats, enzymatic restriction sites.</span>


<span class="sd">&quot;&quot;&quot;</span>

<span class="kn">import</span> <span class="nn">re</span>
<span class="kn">import</span> <span class="nn">itertools</span>
<span class="kn">from</span> <span class="nn">.Location</span> <span class="k">import</span> <span class="n">Location</span>
<span class="kn">from</span> <span class="nn">.biotools</span> <span class="k">import</span> <span class="p">(</span><span class="n">dna_pattern_to_regexpr</span><span class="p">,</span> <span class="n">is_palyndromic</span><span class="p">,</span>
                       <span class="n">reverse_complement</span><span class="p">,</span> <span class="n">NUCLEOTIDE_TO_REGEXPR</span><span class="p">,</span>
                       <span class="n">IUPAC_NOTATION</span><span class="p">)</span>
<span class="kn">from</span> <span class="nn">Bio.Restriction.Restriction_Dictionary</span> <span class="k">import</span> <span class="n">rest_dict</span>

<div class="viewcode-block" id="SequencePattern"><a class="viewcode-back" href="../../ref/core_classes.html#dnachisel.SequencePattern.SequencePattern">[docs]</a><span class="k">class</span> <span class="nc">SequencePattern</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;Pattern/ that will be looked for in a DNA sequence.</span>

<span class="sd">    Use this class for matching regular expression patterns, and</span>
<span class="sd">    DnaNotationPattern for matching explicit sequences or sequences using Ns</span>
<span class="sd">    etc.</span>

<span class="sd">    Example</span>
<span class="sd">    -------</span>
<span class="sd">    &gt;&gt;&gt; expression = &quot;A[ATGC]{3,}&quot;</span>
<span class="sd">    &gt;&gt;&gt; pattern = SequencePattern(expression)</span>
<span class="sd">    &gt;&gt;&gt; constraint = AvoidPattern(pattern)</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>

<span class="sd">    expression</span>
<span class="sd">      Any string or regular expression for matching ATGC nucleotides.</span>
<span class="sd">      Note that multi-nucleotides symbols such as &quot;N&quot; (for A-T-G-C), or &quot;K&quot;</span>
<span class="sd">      are not supported by this class, see DnaNotationPattern instead.</span>

<span class="sd">    size</span>
<span class="sd">      Size of the pattern, in number of characters (if none provided, the size</span>
<span class="sd">      of the ``pattern`` string is used).</span>
<span class="sd">      The ``size`` is used to determine the size of windows when performing</span>
<span class="sd">      local optimization and constraint solving.</span>
<span class="sd">      It can be important to provide the size when the</span>
<span class="sd">      ``pattern`` string provided represents a complex regular expression whose</span>
<span class="sd">      maximal matching size cannot be easily evaluated.</span>

<span class="sd">    name</span>
<span class="sd">      Name of the pattern (will be displayed e.g. when the pattern is printed)</span>

<span class="sd">    in_both_strands</span>
<span class="sd">      Set to True (default) if the pattern should also be looked for on the</span>
<span class="sd">      reverse-complement of sequences.</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">expression</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">in_both_strands</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                 <span class="n">use_lookahead</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
        <span class="k">if</span> <span class="n">size</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">size</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">expression</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">expression</span> <span class="o">=</span> <span class="n">expression</span>
        <span class="k">if</span> <span class="n">use_lookahead</span><span class="p">:</span>
            <span class="n">expression</span> <span class="o">=</span> <span class="s1">&#39;(?=(</span><span class="si">%s</span><span class="s1">))&#39;</span> <span class="o">%</span> <span class="n">expression</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">lookahead_expression</span> <span class="o">=</span> <span class="n">expression</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">size</span> <span class="o">=</span> <span class="n">size</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="n">name</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">in_both_strands</span> <span class="o">=</span> <span class="n">in_both_strands</span>


<div class="viewcode-block" id="SequencePattern.find_matches"><a class="viewcode-back" href="../../ref/core_classes.html#dnachisel.SequencePattern.SequencePattern.find_matches">[docs]</a>    <span class="k">def</span> <span class="nf">find_matches</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">sequence</span><span class="p">,</span> <span class="n">location</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Return the locations where the sequence matches the expression.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>

<span class="sd">        sequence</span>
<span class="sd">          A string of &quot;ATGC...&quot;</span>

<span class="sd">        location</span>
<span class="sd">          Location indicating a segment to which to restrict</span>
<span class="sd">          the search. Only patterns entirely included in the segment will be</span>
<span class="sd">          returned</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>

<span class="sd">        matches</span>
<span class="sd">          A list of the locations of matches, of the form</span>
<span class="sd">          ``[(start1, end1), (start2, end2),...]``.</span>

<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="n">location</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">subsequence</span> <span class="o">=</span> <span class="n">location</span><span class="o">.</span><span class="n">extract_sequence</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span>
            <span class="k">return</span> <span class="p">[</span>
                <span class="n">loc</span> <span class="o">+</span> <span class="n">location</span><span class="o">.</span><span class="n">start</span>
                <span class="k">for</span> <span class="n">loc</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">find_matches</span><span class="p">(</span><span class="n">subsequence</span><span class="p">)</span>
            <span class="p">]</span>
        <span class="n">matches</span> <span class="o">=</span> <span class="p">[</span>
            <span class="p">(</span><span class="n">match</span><span class="o">.</span><span class="n">start</span><span class="p">(),</span> <span class="n">match</span><span class="o">.</span><span class="n">start</span><span class="p">()</span> <span class="o">+</span> <span class="nb">len</span><span class="p">(</span><span class="n">match</span><span class="o">.</span><span class="n">groups</span><span class="p">()[</span><span class="mi">0</span><span class="p">]),</span> <span class="mi">1</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">match</span> <span class="ow">in</span> <span class="n">re</span><span class="o">.</span><span class="n">finditer</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">lookahead_expression</span><span class="p">,</span> <span class="n">sequence</span><span class="p">)</span>
        <span class="p">]</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">in_both_strands</span><span class="p">:</span>
            <span class="n">reverse</span> <span class="o">=</span> <span class="n">reverse_complement</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span>
            <span class="n">L</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span>
            <span class="n">matches</span> <span class="o">+=</span> <span class="p">[</span>
                <span class="p">(</span><span class="n">L</span> <span class="o">-</span> <span class="n">match</span><span class="o">.</span><span class="n">start</span><span class="p">()</span> <span class="o">-</span> <span class="nb">len</span><span class="p">(</span><span class="n">match</span><span class="o">.</span><span class="n">groups</span><span class="p">()[</span><span class="mi">0</span><span class="p">]),</span>
                 <span class="n">L</span> <span class="o">-</span> <span class="n">match</span><span class="o">.</span><span class="n">start</span><span class="p">(),</span> <span class="o">-</span><span class="mi">1</span><span class="p">)</span>
                <span class="k">for</span> <span class="n">match</span> <span class="ow">in</span> <span class="n">re</span><span class="o">.</span><span class="n">finditer</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">lookahead_expression</span><span class="p">,</span> <span class="n">reverse</span><span class="p">)</span>
            <span class="p">]</span>

        <span class="k">return</span> <span class="p">[</span>
            <span class="n">Location</span><span class="p">(</span><span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">strand</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">strand</span> <span class="ow">in</span> <span class="n">matches</span>
        <span class="p">]</span></div>

    <span class="k">def</span> <span class="nf">__str__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">expression</span> <span class="o">+</span> <span class="p">(</span><span class="s2">&quot;&quot;</span> <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">name</span> <span class="ow">is</span> <span class="kc">None</span> <span class="k">else</span>
                                  <span class="s2">&quot; (</span><span class="si">%s</span><span class="s2">)&quot;</span> <span class="o">%</span> <span class="bp">self</span><span class="o">.</span><span class="n">name</span><span class="p">)</span></div>


<div class="viewcode-block" id="DnaNotationPattern"><a class="viewcode-back" href="../../ref/core_classes.html#dnachisel.SequencePattern.DnaNotationPattern">[docs]</a><span class="k">class</span> <span class="nc">DnaNotationPattern</span><span class="p">(</span><span class="n">SequencePattern</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Class for patterns in plain DNA notation: ATTGCCA, GCNNKTA, etc.</span>

<span class="sd">    If the sequence is not palyndromic, the pattern will be looked for in</span>
<span class="sd">    both strands of sequences.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">sequence</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">in_both_strands</span><span class="o">=</span><span class="s1">&#39;auto&#39;</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Initialize&quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">in_both_strands</span> <span class="o">==</span> <span class="s1">&#39;auto&#39;</span><span class="p">:</span>
            <span class="n">in_both_strands</span> <span class="o">=</span> <span class="ow">not</span> <span class="n">is_palyndromic</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span>
        <span class="n">SequencePattern</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span>
            <span class="n">size</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="n">sequence</span><span class="p">),</span>
            <span class="n">expression</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">dna_sequence_to_regexpr</span><span class="p">(</span><span class="n">sequence</span><span class="p">),</span>
            <span class="n">name</span><span class="o">=</span><span class="n">name</span><span class="p">,</span>
            <span class="n">in_both_strands</span><span class="o">=</span><span class="n">in_both_strands</span>
        <span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">sequence</span> <span class="o">=</span> <span class="n">sequence</span>

<div class="viewcode-block" id="DnaNotationPattern.dna_sequence_to_regexpr"><a class="viewcode-back" href="../../ref/core_classes.html#dnachisel.SequencePattern.DnaNotationPattern.dna_sequence_to_regexpr">[docs]</a>    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">dna_sequence_to_regexpr</span><span class="p">(</span><span class="n">sequence</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Return a regular expression to find the pattern in a sequence.&quot;&quot;&quot;</span>

        <span class="n">regexpr</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span>
            <span class="n">NUCLEOTIDE_TO_REGEXPR</span><span class="p">[</span><span class="n">n</span><span class="p">]</span>
            <span class="k">for</span> <span class="n">n</span> <span class="ow">in</span> <span class="n">sequence</span>
        <span class="p">])</span>
        <span class="k">return</span> <span class="n">regexpr</span></div>
        <span class="c1"># The ?= implements &#39;lookahead&#39; which enables to find overlapping</span>
        <span class="c1"># patterns. It is followed by (.{S}) where S is the sequence size</span>
        <span class="c1"># to make sure the full group is selected.</span>
        <span class="c1"># return &#39;(?=(%s))(.\{%d\})&#39; % (regexpr, len(sequence))</span>

<div class="viewcode-block" id="DnaNotationPattern.all_variants"><a class="viewcode-back" href="../../ref/core_classes.html#dnachisel.SequencePattern.DnaNotationPattern.all_variants">[docs]</a>    <span class="k">def</span> <span class="nf">all_variants</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Return all ATGC sequence variants of a sequence&quot;&quot;&quot;</span>
        <span class="k">return</span><span class="p">[</span>
            <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">nucleotides</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">nucleotides</span> <span class="ow">in</span> <span class="n">itertools</span><span class="o">.</span><span class="n">product</span><span class="p">(</span><span class="o">*</span><span class="p">[</span>
                <span class="n">IUPAC_NOTATION</span><span class="p">[</span><span class="n">n</span><span class="p">]</span> <span class="k">for</span> <span class="n">n</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence</span>
            <span class="p">])</span>
        <span class="p">]</span></div>

    <span class="k">def</span> <span class="nf">__repr__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Represent the pattern as PatternType(name) &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence</span> <span class="o">+</span> <span class="p">(</span><span class="s2">&quot;&quot;</span> <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">name</span> <span class="ow">is</span> <span class="kc">None</span> <span class="k">else</span>
                                <span class="s2">&quot; (</span><span class="si">%s</span><span class="s2">)&quot;</span> <span class="o">%</span> <span class="bp">self</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>
    <span class="k">def</span> <span class="nf">__str__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Represent the pattern as PatternType(name) &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence</span> <span class="o">+</span> <span class="p">(</span><span class="s2">&quot;&quot;</span> <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">name</span> <span class="ow">is</span> <span class="kc">None</span> <span class="k">else</span>
                                <span class="s2">&quot; (</span><span class="si">%s</span><span class="s2">)&quot;</span> <span class="o">%</span> <span class="bp">self</span><span class="o">.</span><span class="n">name</span><span class="p">)</span></div>

<span class="c1"># DEFINITION OF COMMON PATTERNS</span>


<div class="viewcode-block" id="homopolymer_pattern"><a class="viewcode-back" href="../../ref/core_classes.html#dnachisel.SequencePattern.homopolymer_pattern">[docs]</a><span class="k">def</span> <span class="nf">homopolymer_pattern</span><span class="p">(</span><span class="n">nucleotides</span><span class="p">,</span> <span class="n">number</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Return a DnaNotationPattern with the sequence of a homopolymer.</span>

<span class="sd">    Examples</span>
<span class="sd">    --------</span>

<span class="sd">    &gt;&gt;&gt; homopolymer(&quot;A&quot;, 6) # returns DnaNotationPattern(&quot;AAAAAA&quot;)</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">nucleotides</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">DnaNotationPattern</span><span class="p">(</span><span class="n">number</span> <span class="o">*</span> <span class="n">nucleotides</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">SequencePattern</span><span class="p">(</span><span class="n">number</span> <span class="o">*</span> <span class="p">(</span><span class="s2">&quot;[</span><span class="si">%s</span><span class="s2">]&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="s2">&quot;|&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">nucleotides</span><span class="p">))),</span>
                               <span class="n">size</span><span class="o">=</span><span class="n">number</span><span class="p">)</span></div>

<div class="viewcode-block" id="enzyme_pattern"><a class="viewcode-back" href="../../ref/core_classes.html#dnachisel.SequencePattern.enzyme_pattern">[docs]</a><span class="k">def</span> <span class="nf">enzyme_pattern</span><span class="p">(</span><span class="n">enzyme_name</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Return a DnaNotationPattern with the sequence of a homopolymer.</span>

<span class="sd">    Examples</span>
<span class="sd">    --------</span>

<span class="sd">    &gt;&gt;&gt; pattern = enzyme_pattern(&quot;BsaI&quot;) # returns DnaNotationPattern(&quot;GGTCTC&quot;)</span>
<span class="sd">    &gt;&gt;&gt; constraint = AvoidPattern(pattern)</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">enzyme_site</span> <span class="o">=</span> <span class="n">rest_dict</span><span class="p">[</span><span class="n">enzyme_name</span><span class="p">][</span><span class="s2">&quot;site&quot;</span><span class="p">]</span>
    <span class="k">return</span> <span class="n">DnaNotationPattern</span><span class="p">(</span><span class="n">enzyme_site</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="n">enzyme_name</span><span class="p">)</span></div>


<div class="viewcode-block" id="repeated_kmers"><a class="viewcode-back" href="../../ref/core_classes.html#dnachisel.SequencePattern.repeated_kmers">[docs]</a><span class="k">def</span> <span class="nf">repeated_kmers</span><span class="p">(</span><span class="n">kmer_size</span><span class="p">,</span> <span class="n">n_repeats</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Return a SequencePattern matching all k-mers repeated n times.</span>

<span class="sd">    Examples</span>
<span class="sd">    --------</span>

<span class="sd">    &gt;&gt;&gt; pattern = repeated_kmers(2, 5) # Result will match e.g. ACACACACAC</span>
<span class="sd">    &gt;&gt;&gt; constraint = AvoidPattern(pattern)</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">return</span> <span class="n">SequencePattern</span><span class="p">(</span>
        <span class="n">size</span><span class="o">=</span><span class="n">kmer_size</span> <span class="o">*</span> <span class="n">n_repeats</span><span class="p">,</span>
        <span class="n">expression</span><span class="o">=</span><span class="sa">r</span><span class="s2">&quot;([ATGC]{</span><span class="si">%d</span><span class="s2">})\1{</span><span class="si">%d</span><span class="s2">}&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">kmer_size</span><span class="p">,</span> <span class="n">n_repeats</span><span class="o">-</span><span class="mi">1</span><span class="p">),</span>
        <span class="n">name</span><span class="o">=</span><span class="s2">&quot;</span><span class="si">%d</span><span class="s2">-repeats </span><span class="si">%d</span><span class="s2">-mers&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">n_repeats</span><span class="p">,</span> <span class="n">kmer_size</span><span class="p">),</span>
        <span class="n">in_both_strands</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>  <span class="c1"># a kmer repeat one strand is also on the other</span>
        <span class="n">use_lookahead</span><span class="o">=</span><span class="kc">False</span>
    <span class="p">)</span></div>
</pre></div>

           </div>
           <div class="articleComments">
            
           </div>
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2017, Edinburgh Genome Foundry.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'../../',
            VERSION:'0.1.0',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true,
            SOURCELINK_SUFFIX: '.txt'
        };
    </script>
      <script type="text/javascript" src="../../_static/jquery.js"></script>
      <script type="text/javascript" src="../../_static/underscore.js"></script>
      <script type="text/javascript" src="../../_static/doctools.js"></script>
      <script type="text/javascript" src="https://cdn.rawgit.com/knsv/mermaid/7.0.0/dist/mermaid.min.js"></script>

  

  
  
    <script type="text/javascript" src="../../_static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
   

</body>
</html>