

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>dnachisel.biotools.biotools &mdash; DnaChisel 0.1.0 documentation</title>
  

  
  
  
  

  

  
  
    

  

  
  
    <link rel="stylesheet" href="../../../_static/css/theme.css" type="text/css" />
  

  
    <link rel="stylesheet" href="../../../_static/css/main.css" type="text/css" />
  
    <link rel="stylesheet" href="https://cdn.rawgit.com/knsv/mermaid/7.0.0/dist/mermaid.css" type="text/css" />
  

  
        <link rel="index" title="Index"
              href="../../../genindex.html"/>
        <link rel="search" title="Search" href="../../../search.html"/>
    <link rel="top" title="DnaChisel 0.1.0 documentation" href="../../../index.html"/>
        <link rel="up" title="Module code" href="../../index.html"/> 

  
  <script src="../../../_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav" role="document">

   
  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="../../../index.html" class="icon icon-home"> DnaChisel
          

          
            
            <img src="../../../_static/html_logo.png" class="logo" />
          
          </a>

          
            
            
              <div class="version">
                0.1.0
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../../../index.html">DnaChisel</a></li>
</ul>
<p class="caption"><span class="caption-text">Reference</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../ref/ref.html">DnaChisel Reference manual</a></li>
</ul>
<p class="caption"><span class="caption-text">Examples</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../examples/plasmid_optimization.html">Plasmid optimization with DnaChisel</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../examples/non_unique_kmers_minimization.html">Minimization of the number of repeated kmers</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../examples/pattern_instertion.html">Pattern insertion in a DNA sequence</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../index.html">DnaChisel</a>
        
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../../index.html">Docs</a> &raquo;</li>
        
          <li><a href="../../index.html">Module code</a> &raquo;</li>
        
      <li>dnachisel.biotools.biotools</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <h1>Source code for dnachisel.biotools.biotools</h1><div class="highlight"><pre>
<span></span><span class="kn">import</span> <span class="nn">tempfile</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">subprocess</span>
<span class="kn">import</span> <span class="nn">time</span>
<span class="kn">from</span> <span class="nn">copy</span> <span class="k">import</span> <span class="n">deepcopy</span>

<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="k">import</span> <span class="n">Seq</span>
<span class="kn">from</span> <span class="nn">Bio.Blast</span> <span class="k">import</span> <span class="n">NCBIXML</span>
<span class="kn">from</span> <span class="nn">Bio.SeqRecord</span> <span class="k">import</span> <span class="n">SeqRecord</span>
<span class="kn">from</span> <span class="nn">Bio.Alphabet</span> <span class="k">import</span> <span class="n">DNAAlphabet</span>
<span class="kn">from</span> <span class="nn">Bio.SeqFeature</span> <span class="k">import</span> <span class="n">SeqFeature</span><span class="p">,</span> <span class="n">FeatureLocation</span>
<span class="kn">from</span> <span class="nn">Bio</span> <span class="k">import</span> <span class="n">SeqIO</span>

<span class="kn">from</span> <span class="nn">.biotables</span> <span class="k">import</span> <span class="n">CODONS_SEQUENCES</span><span class="p">,</span> <span class="n">NUCLEOTIDE_TO_REGEXPR</span>


<div class="viewcode-block" id="complement"><a class="viewcode-back" href="../../../ref/biotools.html#dnachisel.biotools.biotools.complement">[docs]</a><span class="k">def</span> <span class="nf">complement</span><span class="p">(</span><span class="n">dna_sequence</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Return the complement of the DNA sequence.</span>

<span class="sd">    For instance ``complement(&quot;ATGCCG&quot;)`` returns ``&quot;TACGGC&quot;``.</span>

<span class="sd">    Uses BioPython for speed.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="nb">str</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="n">dna_sequence</span><span class="p">)</span><span class="o">.</span><span class="n">complement</span><span class="p">())</span></div>


<div class="viewcode-block" id="reverse_complement"><a class="viewcode-back" href="../../../ref/biotools.html#dnachisel.biotools.biotools.reverse_complement">[docs]</a><span class="k">def</span> <span class="nf">reverse_complement</span><span class="p">(</span><span class="n">sequence</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Return the reverse-complement of the DNA sequence.</span>

<span class="sd">    For instance ``complement(&quot;ATGCCG&quot;)`` returns ``&quot;GCCGTA&quot;``.</span>

<span class="sd">    Uses BioPython for speed.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">complement</span><span class="p">(</span><span class="n">sequence</span><span class="p">)[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span></div>


<div class="viewcode-block" id="is_palyndromic"><a class="viewcode-back" href="../../../ref/biotools.html#dnachisel.biotools.biotools.is_palyndromic">[docs]</a><span class="k">def</span> <span class="nf">is_palyndromic</span><span class="p">(</span><span class="n">dna_sequence</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Return True if the DNA sequence is equal to its reverse complement.&quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">reverse_complement</span><span class="p">(</span><span class="n">dna_sequence</span><span class="p">)</span> <span class="o">==</span> <span class="n">dna_sequence</span></div>


<div class="viewcode-block" id="random_dna_sequence"><a class="viewcode-back" href="../../../ref/biotools.html#dnachisel.biotools.biotools.random_dna_sequence">[docs]</a><span class="k">def</span> <span class="nf">random_dna_sequence</span><span class="p">(</span><span class="n">length</span><span class="p">,</span> <span class="n">probas</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">seed</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Return a random DNA sequence (&quot;ATGGCGT...&quot;) with the specified length.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>

<span class="sd">    length</span>
<span class="sd">      Length of the DNA sequence.</span>

<span class="sd">    proba</span>
<span class="sd">      Frequencies for the different nucleotides, for instance</span>
<span class="sd">      ``probas={&quot;A&quot;:0.2, &quot;T&quot;:0.3, &quot;G&quot;:0.3, &quot;C&quot;:0.2}``.</span>
<span class="sd">      If not specified, all nucleotides are equiprobable (p=0.25).</span>

<span class="sd">    seed</span>
<span class="sd">      The seed to feed to the random number generator. When a seed is provided</span>
<span class="sd">      the random results depend deterministically on the seed, thus enabling</span>
<span class="sd">      reproducibility</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">seed</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">seed</span><span class="p">(</span><span class="n">seed</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">probas</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">sequence</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="s2">&quot;ATCG&quot;</span><span class="p">),</span> <span class="n">length</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">bases</span><span class="p">,</span> <span class="n">probas</span> <span class="o">=</span> <span class="nb">zip</span><span class="p">(</span><span class="o">*</span><span class="n">probas</span><span class="o">.</span><span class="n">items</span><span class="p">())</span>
        <span class="n">sequence</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="n">bases</span><span class="p">,</span> <span class="n">length</span><span class="p">,</span> <span class="n">p</span><span class="o">=</span><span class="n">probas</span><span class="p">)</span>
    <span class="k">return</span> <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span></div>


<span class="k">def</span> <span class="nf">load_record</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="n">linear</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s2">&quot;unnamed&quot;</span><span class="p">):</span>
    <span class="k">if</span> <span class="n">filename</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span><span class="o">.</span><span class="n">endswith</span><span class="p">((</span><span class="s2">&quot;gb&quot;</span><span class="p">,</span> <span class="s2">&quot;gbk&quot;</span><span class="p">)):</span>
        <span class="n">record</span> <span class="o">=</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="s2">&quot;genbank&quot;</span><span class="p">)</span>
    <span class="k">elif</span> <span class="n">filename</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span><span class="o">.</span><span class="n">endswith</span><span class="p">((</span><span class="s1">&#39;fa&#39;</span><span class="p">,</span> <span class="s1">&#39;fasta&#39;</span><span class="p">)):</span>
        <span class="n">record</span> <span class="o">=</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;Unknown format for file: </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="n">filename</span><span class="p">)</span>
    <span class="n">record</span><span class="o">.</span><span class="n">linear</span> <span class="o">=</span> <span class="n">linear</span>
    <span class="n">record</span><span class="o">.</span><span class="n">id</span> <span class="o">=</span> <span class="n">name</span>
    <span class="n">record</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="n">name</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot; &quot;</span><span class="p">,</span> <span class="s2">&quot;_&quot;</span><span class="p">)[:</span><span class="mi">20</span><span class="p">]</span>
    <span class="k">return</span> <span class="n">record</span>

<div class="viewcode-block" id="random_protein_sequence"><a class="viewcode-back" href="../../../ref/biotools.html#dnachisel.biotools.biotools.random_protein_sequence">[docs]</a><span class="k">def</span> <span class="nf">random_protein_sequence</span><span class="p">(</span><span class="n">length</span><span class="p">,</span> <span class="n">seed</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Return a random protein sequence &quot;MNQTW...YL*&quot; of the specified length.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>

<span class="sd">    length</span>
<span class="sd">      Length of the protein sequence (in number of amino-acids). Note that the</span>
<span class="sd">      sequence will always start with ``&quot;M&quot;`` and end with a stop codon ``&quot;*&quot;``</span>
<span class="sd">      with (length-2) random amino-acids in the middle</span>

<span class="sd">    seed</span>
<span class="sd">      The seed to feed to the random number generator. When a seed is provided</span>
<span class="sd">      the random results depend deterministically on the seed, thus enabling</span>
<span class="sd">      reproducibility</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">seed</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">seed</span><span class="p">(</span><span class="n">seed</span><span class="p">)</span>

    <span class="n">aa_list</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="s1">&#39;ACEDGFIHKLNQPSRTWVY&#39;</span><span class="p">)</span>
    <span class="n">aa_choices</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="n">aa_list</span><span class="p">,</span> <span class="n">length</span> <span class="o">-</span> <span class="mi">2</span><span class="p">)</span>
    <span class="k">return</span> <span class="s2">&quot;M&quot;</span> <span class="o">+</span> <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">aa_choices</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;*&quot;</span></div>


<div class="viewcode-block" id="reverse_translate"><a class="viewcode-back" href="../../../ref/biotools.html#dnachisel.biotools.biotools.reverse_translate">[docs]</a><span class="k">def</span> <span class="nf">reverse_translate</span><span class="p">(</span><span class="n">protein_sequence</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Return a DNA sequence which translates to the provided protein sequence</span>

<span class="sd">    Note: at the moment, the first valid codon found is used for each</span>
<span class="sd">    amino-acid (so it is deterministic but no codon-optimization is done).</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span>
        <span class="n">CODONS_SEQUENCES</span><span class="p">[</span><span class="n">aa</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
        <span class="k">for</span> <span class="n">aa</span> <span class="ow">in</span> <span class="n">protein_sequence</span>
    <span class="p">])</span></div>


<div class="viewcode-block" id="translate"><a class="viewcode-back" href="../../../ref/biotools.html#dnachisel.biotools.biotools.translate">[docs]</a><span class="k">def</span> <span class="nf">translate</span><span class="p">(</span><span class="n">dna_sequence</span><span class="p">,</span> <span class="n">translation_table</span><span class="o">=</span><span class="s2">&quot;Bacterial&quot;</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Translate the DNA sequence into an amino-acids sequence &quot;MLKYQT...&quot;.</span>
<span class="sd">    If ``translation_table`` is the name or number of  NCBI genetic table,</span>
<span class="sd">    Biopython will be used. See here for options:</span>

<span class="sd">    http://biopython.org/DIST/docs/tutorial/Tutorial.html#htoc25</span>

<span class="sd">    ``translation_table`` can also be a dictionnary of the form</span>
<span class="sd">    ``{&quot;ATT&quot;: &quot;M&quot;, &quot;CTC&quot;: &quot;X&quot;, etc.}`` for more exotic translation tables</span>


<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">translation_table</span><span class="p">,</span> <span class="nb">dict</span><span class="p">):</span>
        <span class="k">return</span> <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span>
            <span class="n">translation_table</span><span class="p">[</span><span class="n">dna_sequence</span><span class="p">[</span><span class="n">i</span><span class="p">:</span><span class="n">i</span> <span class="o">+</span> <span class="mi">3</span><span class="p">]]</span>
            <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">dna_sequence</span><span class="p">),</span> <span class="mi">3</span><span class="p">)</span>
        <span class="p">])</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">return</span> <span class="nb">str</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="n">dna_sequence</span><span class="p">)</span><span class="o">.</span><span class="n">translate</span><span class="p">(</span><span class="n">table</span><span class="o">=</span><span class="n">translation_table</span><span class="p">))</span></div>


<div class="viewcode-block" id="dna_pattern_to_regexpr"><a class="viewcode-back" href="../../../ref/biotools.html#dnachisel.biotools.biotools.dna_pattern_to_regexpr">[docs]</a><span class="k">def</span> <span class="nf">dna_pattern_to_regexpr</span><span class="p">(</span><span class="n">dna_pattern</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Return a regular expression pattern for the provided DNA pattern</span>

<span class="sd">    For instance ``dna_pattern_to_regexpr(&#39;ATTNN&#39;)`` returns</span>
<span class="sd">    ``&quot;ATT[A|T|G|C][A|T|G|C]&quot;``.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span>
        <span class="n">NUCLEOTIDE_TO_REGEXPR</span><span class="p">[</span><span class="n">n</span><span class="p">]</span>
        <span class="k">for</span> <span class="n">n</span> <span class="ow">in</span> <span class="n">dna_pattern</span>
    <span class="p">])</span></div>


<div class="viewcode-block" id="windows_overlap"><a class="viewcode-back" href="../../../ref/biotools.html#dnachisel.biotools.biotools.windows_overlap">[docs]</a><span class="k">def</span> <span class="nf">windows_overlap</span><span class="p">(</span><span class="n">window1</span><span class="p">,</span> <span class="n">window2</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Return the overlap span between two windows.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>

<span class="sd">    window1, window2</span>
<span class="sd">      Each window is a couple of the form (start, end) indicating the range of</span>
<span class="sd">      a segment of integers.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>

<span class="sd">    None</span>
<span class="sd">      In case the two windows do not overlap.</span>
<span class="sd">    [start, end]</span>
<span class="sd">      The coordinates of the overlap segment if there is one.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">start1</span><span class="p">,</span> <span class="n">end1</span> <span class="o">=</span> <span class="n">window1</span>
    <span class="n">start2</span><span class="p">,</span> <span class="n">end2</span> <span class="o">=</span> <span class="n">window2</span>

    <span class="k">if</span> <span class="n">start2</span> <span class="o">&lt;</span> <span class="n">start1</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">windows_overlap</span><span class="p">(</span><span class="n">window2</span><span class="p">,</span> <span class="n">window1</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">start1</span> <span class="o">&lt;=</span> <span class="n">start2</span> <span class="o">&lt;=</span> <span class="n">end1</span><span class="p">:</span>
        <span class="k">return</span> <span class="p">[</span><span class="n">start2</span><span class="p">,</span> <span class="nb">min</span><span class="p">(</span><span class="n">end1</span><span class="p">,</span> <span class="n">end2</span><span class="p">)]</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">return</span> <span class="kc">None</span></div>


<div class="viewcode-block" id="read_fasta"><a class="viewcode-back" href="../../../ref/biotools.html#dnachisel.biotools.biotools.read_fasta">[docs]</a><span class="k">def</span> <span class="nf">read_fasta</span><span class="p">(</span><span class="n">filename</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Read A sequence in a FASTA file with Biopython.&quot;&quot;&quot;</span>
    <span class="kn">import</span> <span class="nn">Bio.SeqIO</span> <span class="k">as</span> <span class="nn">seqio</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">filename</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
        <span class="k">return</span> <span class="nb">str</span><span class="p">(</span><span class="n">seqio</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">f</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">seq</span><span class="p">)</span></div>


<div class="viewcode-block" id="gc_content"><a class="viewcode-back" href="../../../ref/biotools.html#dnachisel.biotools.biotools.gc_content">[docs]</a><span class="k">def</span> <span class="nf">gc_content</span><span class="p">(</span><span class="n">sequence</span><span class="p">,</span> <span class="n">window_size</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Compute global or local GC content.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>

<span class="sd">    sequence</span>
<span class="sd">      An ATGC DNA sequence (upper case!)</span>

<span class="sd">    window_size</span>
<span class="sd">      If provided, the local GC content for the different sliding windows of</span>
<span class="sd">      this size is returned, else the global GC content is returned.</span>

<span class="sd">    Returns</span>
<span class="sd">    --------</span>

<span class="sd">      A number between 0 and 1 indication the proportion</span>
<span class="sd">      of GC content. If window_size is provided, returns</span>
<span class="sd">      a list of len(sequence)-window_size values indicating</span>
<span class="sd">      the local GC contents (sliding-window method). The i-th value</span>
<span class="sd">      indicates the GC content in the window [i, i+window_size]</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># The code is a little cryptic but speed gain is 300x</span>
    <span class="c1"># compared with pure-python string operations</span>

    <span class="n">arr</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">fromstring</span><span class="p">(</span><span class="n">sequence</span> <span class="o">+</span> <span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="s2">&quot;uint8&quot;</span><span class="p">)</span>
    <span class="n">arr_GCs</span> <span class="o">=</span> <span class="p">(</span><span class="n">arr</span> <span class="o">==</span> <span class="mi">71</span><span class="p">)</span> <span class="o">|</span> <span class="p">(</span><span class="n">arr</span> <span class="o">==</span> <span class="mi">67</span><span class="p">)</span>  <span class="c1"># 67=C, 71=G</span>

    <span class="k">if</span> <span class="n">window_size</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">return</span> <span class="mf">1.0</span> <span class="o">*</span> <span class="n">arr_GCs</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span> <span class="o">/</span> <span class="nb">len</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">cs</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">cumsum</span><span class="p">(</span><span class="n">arr_GCs</span><span class="p">)</span>
        <span class="n">a</span> <span class="o">=</span> <span class="n">cs</span><span class="p">[</span><span class="n">window_size</span> <span class="o">-</span> <span class="mi">1</span><span class="p">:]</span>
        <span class="n">b</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">hstack</span><span class="p">([[</span><span class="mi">0</span><span class="p">],</span> <span class="n">cs</span><span class="p">[:</span><span class="o">-</span><span class="n">window_size</span><span class="p">]])</span>
        <span class="k">return</span> <span class="mf">1.0</span> <span class="o">*</span> <span class="p">(</span><span class="n">a</span> <span class="o">-</span> <span class="n">b</span><span class="p">)</span> <span class="o">/</span> <span class="n">window_size</span></div>


<span class="c1"># def find_orfs(sequence, minsize=300):</span>
<span class="c1">#     &quot;&quot;&quot;Return the list of (start, end) of all orfs in a sequence</span>
<span class="c1">#</span>
<span class="c1">#     TODO: This seems to be junk code from a former project. complete it</span>
<span class="c1">#     or bin it.</span>
<span class="c1">#</span>
<span class="c1">#     &quot;&quot;&quot;</span>
<span class="c1">#     import tempfile</span>
<span class="c1">#     import os</span>
<span class="c1">#     import subprocess as sp</span>
<span class="c1">#     import regex</span>
<span class="c1">#     input_temp = tempfile.mkstemp(suffix=&quot;.fa&quot;)[1]</span>
<span class="c1">#     self.to_fasta(filename=input_temp)</span>
<span class="c1">#     outfile = tempfile.mkstemp(suffix=&quot;.seq&quot;)[1]</span>
<span class="c1">#     proc = sp.Popen([&quot;getorf&quot;, &quot;-minsize&quot;, &quot;%d&quot; % minsize,</span>
<span class="c1">#                      &quot;-sequence&quot;, input_temp, &quot;-outseq&quot;, outfile])</span>
<span class="c1">#     proc.wait()</span>
<span class="c1">#     with open(outfile, &#39;r&#39;) as f:</span>
<span class="c1">#         result = f.read()</span>
<span class="c1">#</span>
<span class="c1">#     os.remove(outfile)</span>
<span class="c1">#     os.remove(input_temp)</span>
<span class="c1">#     orfs_coordinates = regex.findall(&quot;&gt;(\S+) \[(\d+) - (\d+)\]&quot;,</span>
<span class="c1">#                                      result)</span>
<span class="c1">#     return [</span>
<span class="c1">#         (int(start), int(stop), (+1 if int(start) &lt; int(stop) else -1))</span>
<span class="c1">#         for (_, start, stop) in orfs_coordinates</span>
<span class="c1">#     ]</span>


<div class="viewcode-block" id="blast_sequence"><a class="viewcode-back" href="../../../ref/biotools.html#dnachisel.biotools.biotools.blast_sequence">[docs]</a><span class="k">def</span> <span class="nf">blast_sequence</span><span class="p">(</span><span class="n">sequence</span><span class="p">,</span> <span class="n">blast_db</span><span class="p">,</span> <span class="n">word_size</span><span class="o">=</span><span class="mi">4</span><span class="p">,</span> <span class="n">perc_identity</span><span class="o">=</span><span class="mi">80</span><span class="p">,</span>
                   <span class="n">num_alignments</span><span class="o">=</span><span class="mi">1000</span><span class="p">,</span> <span class="n">ungapped</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">num_threads</span><span class="o">=</span><span class="mi">3</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Return a Biopython BLAST record of the given sequence BLASTed</span>
<span class="sd">    against the provided database.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>

<span class="sd">    sequence</span>
<span class="sd">      An ATGC sequence</span>

<span class="sd">    Examples</span>
<span class="sd">    --------</span>

<span class="sd">    &gt;&gt;&gt; blast_record = blast_sequence(&quot;ATTGTGCGTGTGTGCGT&quot;, &quot;blastdb/ecoli&quot;)</span>
<span class="sd">    &gt;&gt;&gt; for alignment in blast_record.alignments:</span>
<span class="sd">    &gt;&gt;&gt;     for hit in alignment.hsps:</span>
<span class="sd">    &gt;&gt;&gt;         print (hit.identities)</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">xml_file</span><span class="p">,</span> <span class="n">xml_name</span> <span class="o">=</span> <span class="n">tempfile</span><span class="o">.</span><span class="n">mkstemp</span><span class="p">(</span><span class="s2">&quot;.xml&quot;</span><span class="p">)</span>
    <span class="n">fasta_file</span><span class="p">,</span> <span class="n">fasta_name</span> <span class="o">=</span> <span class="n">tempfile</span><span class="o">.</span><span class="n">mkstemp</span><span class="p">(</span><span class="s2">&quot;.fa&quot;</span><span class="p">)</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">fasta_name</span><span class="p">,</span> <span class="s2">&quot;w+&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
        <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;&gt;seq</span><span class="se">\n</span><span class="s2">&quot;</span> <span class="o">+</span> <span class="n">sequence</span><span class="p">)</span>

    <span class="n">p</span> <span class="o">=</span> <span class="n">subprocess</span><span class="o">.</span><span class="n">Popen</span><span class="p">(</span>
        <span class="p">[</span>
            <span class="s2">&quot;blastn&quot;</span><span class="p">,</span> <span class="s2">&quot;-out&quot;</span><span class="p">,</span> <span class="n">xml_name</span><span class="p">,</span>
            <span class="s2">&quot;-outfmt&quot;</span><span class="p">,</span> <span class="s2">&quot;5&quot;</span><span class="p">,</span>
            <span class="s2">&quot;-num_alignments&quot;</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">num_alignments</span><span class="p">),</span>
            <span class="s2">&quot;-query&quot;</span><span class="p">,</span> <span class="n">fasta_name</span><span class="p">,</span>
            <span class="s2">&quot;-db&quot;</span><span class="p">,</span> <span class="n">blast_db</span><span class="p">,</span>
            <span class="s2">&quot;-word_size&quot;</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">word_size</span><span class="p">),</span>
            <span class="s2">&quot;-num_threads&quot;</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">num_threads</span><span class="p">),</span>
            <span class="s2">&quot;-perc_identity&quot;</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">perc_identity</span><span class="p">)</span>
        <span class="p">]</span> <span class="o">+</span> <span class="p">([</span><span class="s2">&quot;-ungapped&quot;</span><span class="p">]</span> <span class="k">if</span> <span class="n">ungapped</span> <span class="k">else</span> <span class="p">[]),</span>
        <span class="n">close_fds</span><span class="o">=</span><span class="kc">True</span>
    <span class="p">)</span>
    <span class="n">p</span><span class="o">.</span><span class="n">communicate</span><span class="p">()</span>
    <span class="n">p</span><span class="o">.</span><span class="n">wait</span><span class="p">()</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">3</span><span class="p">):</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">xml_name</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
                <span class="n">blast_record</span> <span class="o">=</span> <span class="n">NCBIXML</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>
            <span class="k">break</span>
        <span class="k">except</span> <span class="ne">ValueError</span><span class="p">:</span>
            <span class="n">time</span><span class="o">.</span><span class="n">sleep</span><span class="p">(</span><span class="mf">0.1</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Problem reading the blast record.&quot;</span><span class="p">)</span>

    <span class="n">os</span><span class="o">.</span><span class="n">fdopen</span><span class="p">(</span><span class="n">xml_file</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
    <span class="n">os</span><span class="o">.</span><span class="n">fdopen</span><span class="p">(</span><span class="n">fasta_file</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

    <span class="k">return</span> <span class="n">blast_record</span></div>


<div class="viewcode-block" id="subdivide_window"><a class="viewcode-back" href="../../../ref/biotools.html#dnachisel.biotools.biotools.subdivide_window">[docs]</a><span class="k">def</span> <span class="nf">subdivide_window</span><span class="p">(</span><span class="n">window</span><span class="p">,</span> <span class="n">max_span</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Subdivide a window (start, end) into windows of size &lt; max_span</span>
<span class="sd">    (start, i_1), (i_1, i_2), ... (i_n, end)&quot;&quot;&quot;</span>
    <span class="n">start</span><span class="p">,</span> <span class="n">end</span> <span class="o">=</span> <span class="n">window</span>
    <span class="n">inds</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">max_span</span><span class="p">))</span> <span class="o">+</span> <span class="p">[</span><span class="n">end</span><span class="p">]</span>
    <span class="k">return</span> <span class="nb">zip</span><span class="p">(</span><span class="n">inds</span><span class="p">,</span> <span class="n">inds</span><span class="p">[</span><span class="mi">1</span><span class="p">:])</span></div>


<div class="viewcode-block" id="change_biopython_record_sequence"><a class="viewcode-back" href="../../../ref/biotools.html#dnachisel.biotools.biotools.change_biopython_record_sequence">[docs]</a><span class="k">def</span> <span class="nf">change_biopython_record_sequence</span><span class="p">(</span><span class="n">record</span><span class="p">,</span> <span class="n">new_seq</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Return a version of the record with the sequence set to new_seq&quot;&quot;&quot;</span>
    <span class="n">new_record</span> <span class="o">=</span> <span class="n">deepcopy</span><span class="p">(</span><span class="n">record</span><span class="p">)</span>
    <span class="n">new_record</span><span class="o">.</span><span class="n">seq</span> <span class="o">=</span> <span class="n">Seq</span><span class="p">(</span><span class="n">new_seq</span><span class="p">,</span> <span class="n">alphabet</span><span class="o">=</span><span class="n">DNAAlphabet</span><span class="p">())</span>
    <span class="k">return</span> <span class="n">new_record</span></div>


<div class="viewcode-block" id="sequence_to_biopython_record"><a class="viewcode-back" href="../../../ref/biotools.html#dnachisel.biotools.biotools.sequence_to_biopython_record">[docs]</a><span class="k">def</span> <span class="nf">sequence_to_biopython_record</span><span class="p">(</span><span class="n">sequence</span><span class="p">,</span> <span class="nb">id</span><span class="o">=</span><span class="s1">&#39;&lt;unknown id&gt;&#39;</span><span class="p">,</span>
                                 <span class="n">name</span><span class="o">=</span><span class="s1">&#39;&lt;unknown name&gt;&#39;</span><span class="p">,</span> <span class="n">features</span><span class="o">=</span><span class="p">()):</span>
    <span class="sd">&quot;&quot;&quot;Return a SeqRecord of the sequence, ready to be Genbanked.&quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">SeqRecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="n">sequence</span><span class="p">,</span> <span class="n">alphabet</span><span class="o">=</span><span class="n">DNAAlphabet</span><span class="p">()),</span>
                     <span class="nb">id</span><span class="o">=</span><span class="nb">id</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="n">name</span><span class="p">,</span> <span class="n">features</span><span class="o">=</span><span class="nb">list</span><span class="p">(</span><span class="n">features</span><span class="p">))</span></div>


<div class="viewcode-block" id="find_specification_in_feature"><a class="viewcode-back" href="../../../ref/biotools.html#dnachisel.biotools.biotools.find_specification_in_feature">[docs]</a><span class="k">def</span> <span class="nf">find_specification_in_feature</span><span class="p">(</span><span class="n">feature</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Analyse a Biopython feature to find a DnaChisel Specification in it.</span>

<span class="sd">    The specification should start with either &quot;@&quot; or &quot;~&quot;, in the feature&#39;s</span>
<span class="sd">    field &quot;label&quot; or &quot;note&quot;.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">for</span> <span class="n">labelfield</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;label&quot;</span><span class="p">,</span> <span class="s2">&quot;note&quot;</span><span class="p">]:</span>
        <span class="k">if</span> <span class="n">labelfield</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">feature</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">:</span>
            <span class="k">continue</span>
        <span class="n">potential_label</span> <span class="o">=</span> <span class="n">feature</span><span class="o">.</span><span class="n">qualifiers</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">labelfield</span><span class="p">,</span> <span class="s2">&quot;_&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">potential_label</span><span class="p">,</span> <span class="nb">list</span><span class="p">):</span>
            <span class="n">potential_label</span> <span class="o">=</span> <span class="n">potential_label</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="k">if</span> <span class="n">potential_label</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="ow">in</span> <span class="s2">&quot;@~&quot;</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">potential_label</span>
    <span class="k">return</span> <span class="kc">None</span></div>


<div class="viewcode-block" id="crop_record"><a class="viewcode-back" href="../../../ref/biotools.html#dnachisel.biotools.biotools.crop_record">[docs]</a><span class="k">def</span> <span class="nf">crop_record</span><span class="p">(</span><span class="n">record</span><span class="p">,</span> <span class="n">crop_start</span><span class="p">,</span> <span class="n">crop_end</span><span class="p">,</span> <span class="n">features_suffix</span><span class="o">=</span><span class="s2">&quot; (part)&quot;</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Return the cropped record with possibly cropped features.</span>

<span class="sd">    Note that this differs from ``record[start:end]`` in that in the latter</span>
<span class="sd">    expression, cropped features are discarded.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>

<span class="sd">    record</span>
<span class="sd">      A Biopython record</span>

<span class="sd">    crop_start, crop_end</span>
<span class="sd">      Start and end of the segment to be cropped.</span>

<span class="sd">    features_suffix</span>
<span class="sd">      All cropped features will have their label appended with this suffix.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">features</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">feature</span> <span class="ow">in</span> <span class="n">record</span><span class="o">.</span><span class="n">features</span><span class="p">:</span>
        <span class="n">start</span><span class="p">,</span> <span class="n">end</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">([</span><span class="n">feature</span><span class="o">.</span><span class="n">location</span><span class="o">.</span><span class="n">start</span><span class="p">,</span> <span class="n">feature</span><span class="o">.</span><span class="n">location</span><span class="o">.</span><span class="n">end</span><span class="p">])</span>
        <span class="n">new_start</span><span class="p">,</span> <span class="n">new_end</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="n">start</span><span class="p">,</span> <span class="n">crop_start</span><span class="p">),</span> <span class="nb">min</span><span class="p">(</span><span class="n">end</span><span class="p">,</span> <span class="n">crop_end</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">new_end</span> <span class="o">&lt;=</span> <span class="n">new_start</span><span class="p">:</span>
            <span class="k">continue</span>
        <span class="n">new_start</span><span class="p">,</span> <span class="n">new_end</span> <span class="o">=</span> <span class="n">new_start</span> <span class="o">-</span> <span class="n">crop_start</span><span class="p">,</span> <span class="n">new_end</span> <span class="o">-</span> <span class="n">crop_start</span>

        <span class="n">feature</span> <span class="o">=</span> <span class="n">deepcopy</span><span class="p">(</span><span class="n">feature</span><span class="p">)</span>
        <span class="n">feature</span><span class="o">.</span><span class="n">location</span> <span class="o">=</span> <span class="n">FeatureLocation</span><span class="p">(</span><span class="n">new_start</span><span class="p">,</span> <span class="n">new_end</span><span class="p">,</span>
                                           <span class="n">feature</span><span class="o">.</span><span class="n">location</span><span class="o">.</span><span class="n">strand</span><span class="p">)</span>
        <span class="n">label</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">feature</span><span class="o">.</span><span class="n">qualifiers</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;label&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">))</span>
        <span class="n">feature</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;label&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">label</span> <span class="o">+</span> <span class="n">features_suffix</span>
        <span class="n">features</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">feature</span><span class="p">)</span>

    <span class="n">new_record</span> <span class="o">=</span> <span class="n">record</span><span class="p">[</span><span class="n">crop_start</span><span class="p">:</span> <span class="n">crop_end</span><span class="p">]</span>
    <span class="n">new_record</span><span class="o">.</span><span class="n">features</span> <span class="o">=</span> <span class="n">features</span>
    <span class="k">return</span> <span class="n">new_record</span></div>


<div class="viewcode-block" id="sequences_differences_array"><a class="viewcode-back" href="../../../ref/biotools.html#dnachisel.biotools.biotools.sequences_differences_array">[docs]</a><span class="k">def</span> <span class="nf">sequences_differences_array</span><span class="p">(</span><span class="n">seq1</span><span class="p">,</span> <span class="n">seq2</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Return an array [0, 0, 1, 0, ...] with 1s for sequence differences.</span>

<span class="sd">    seq1, seq2 should both be ATGC strings.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">seq1</span><span class="p">)</span> <span class="o">!=</span> <span class="nb">len</span><span class="p">(</span><span class="n">seq2</span><span class="p">):</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Only use on same-size sequences&quot;</span><span class="p">)</span>
    <span class="n">arr1</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">fromstring</span><span class="p">(</span><span class="n">seq1</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="s2">&quot;uint8&quot;</span><span class="p">)</span>
    <span class="n">arr2</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">fromstring</span><span class="p">(</span><span class="n">seq2</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="s2">&quot;uint8&quot;</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">arr1</span> <span class="o">!=</span> <span class="n">arr2</span></div>


<div class="viewcode-block" id="sequences_differences"><a class="viewcode-back" href="../../../ref/biotools.html#dnachisel.biotools.biotools.sequences_differences">[docs]</a><span class="k">def</span> <span class="nf">sequences_differences</span><span class="p">(</span><span class="n">seq1</span><span class="p">,</span> <span class="n">seq2</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Return the number of nucleotides that differ in the two sequences.</span>

<span class="sd">    seq1, seq2 should be strings of DNA sequences e.g. &quot;ATGCTGTGC&quot;</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">sequences_differences_array</span><span class="p">(</span><span class="n">seq1</span><span class="p">,</span> <span class="n">seq2</span><span class="p">)</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span></div>


<div class="viewcode-block" id="sequences_differences_segments"><a class="viewcode-back" href="../../../ref/biotools.html#dnachisel.biotools.biotools.sequences_differences_segments">[docs]</a><span class="k">def</span> <span class="nf">sequences_differences_segments</span><span class="p">(</span><span class="n">seq1</span><span class="p">,</span> <span class="n">seq2</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Return the list of segments on which sequence seq1 differs from seq2.</span>

<span class="sd">    The list is of the form [(start1, end1), (start2, end2), etc.]</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>

<span class="sd">    seq1, seq2</span>
<span class="sd">      ATGC sequences to be compared</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">arr</span> <span class="o">=</span> <span class="mi">1</span> <span class="o">*</span> <span class="n">sequences_differences_array</span><span class="p">(</span><span class="n">seq1</span><span class="p">,</span> <span class="n">seq2</span><span class="p">)</span>
    <span class="n">diffs</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">diff</span><span class="p">([</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="nb">list</span><span class="p">(</span><span class="n">arr</span><span class="p">)</span> <span class="o">+</span> <span class="p">[</span><span class="mi">0</span><span class="p">])</span><span class="o">.</span><span class="n">nonzero</span><span class="p">()[</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">half</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">diffs</span><span class="p">)</span> <span class="o">/</span> <span class="mi">2</span><span class="p">)</span>
    <span class="k">return</span> <span class="p">[(</span><span class="n">diffs</span><span class="p">[</span><span class="mi">2</span> <span class="o">*</span> <span class="n">i</span><span class="p">],</span> <span class="n">diffs</span><span class="p">[</span><span class="mi">2</span> <span class="o">*</span> <span class="n">i</span> <span class="o">+</span> <span class="mi">1</span><span class="p">])</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">half</span><span class="p">)]</span></div>

<div class="viewcode-block" id="group_nearby_indices"><a class="viewcode-back" href="../../../ref/biotools.html#dnachisel.biotools.biotools.group_nearby_indices">[docs]</a><span class="k">def</span> <span class="nf">group_nearby_indices</span><span class="p">(</span><span class="n">indices</span><span class="p">,</span> <span class="n">max_gap</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">max_group_spread</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Return a list of groups of the different indices.</span>

<span class="sd">    Indices are considered from smaller to larger and placed into groups</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    max_gap</span>
<span class="sd">      Maximal allowed difference between two consecutive numbers of a group</span>

<span class="sd">    max_group_spread</span>
<span class="sd">      Maximal allowed difference between the smallest and largest elements</span>
<span class="sd">      of a group.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">indices</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
        <span class="k">return</span> <span class="p">[]</span>
    <span class="n">indices</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">indices</span><span class="p">)</span>
    <span class="n">current_group</span> <span class="o">=</span> <span class="p">[</span><span class="n">indices</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span>
    <span class="n">groups</span> <span class="o">=</span> <span class="p">[</span><span class="n">current_group</span><span class="p">]</span>
    <span class="k">for</span> <span class="n">ind</span> <span class="ow">in</span> <span class="n">indices</span><span class="p">[</span><span class="mi">1</span><span class="p">:]:</span>
        <span class="n">gap_small_enough</span> <span class="o">=</span> <span class="p">((</span><span class="n">max_gap</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">)</span> <span class="ow">or</span>
                            <span class="p">(</span><span class="n">ind</span> <span class="o">-</span> <span class="n">current_group</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">&lt;</span> <span class="n">max_gap</span><span class="p">))</span>
        <span class="n">spread_small_enough</span> <span class="o">=</span> <span class="p">((</span><span class="n">max_group_spread</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">)</span> <span class="ow">or</span>
                               <span class="p">(</span><span class="n">ind</span> <span class="o">-</span> <span class="n">current_group</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">&lt;</span> <span class="n">max_group_spread</span><span class="p">))</span>
        <span class="k">if</span> <span class="n">gap_small_enough</span> <span class="ow">and</span> <span class="n">spread_small_enough</span><span class="p">:</span>
            <span class="n">current_group</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">ind</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">current_group</span> <span class="o">=</span> <span class="p">[</span><span class="n">ind</span><span class="p">]</span>
            <span class="n">groups</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">current_group</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">groups</span></div>

<div class="viewcode-block" id="group_nearby_segments"><a class="viewcode-back" href="../../../ref/biotools.html#dnachisel.biotools.biotools.group_nearby_segments">[docs]</a><span class="k">def</span> <span class="nf">group_nearby_segments</span><span class="p">(</span><span class="n">segments</span><span class="p">,</span> <span class="n">max_start_gap</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">max_start_spread</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Return a list of groups of the different indices.</span>

<span class="sd">    Indices are considered from smaller to larger and placed into groups</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    max_gap</span>
<span class="sd">      Maximal allowed difference between two consecutive numbers of a group</span>

<span class="sd">    max_group_spread</span>
<span class="sd">      Maximal allowed difference between the smallest and largest elements</span>
<span class="sd">      of a group.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">segments</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
        <span class="k">return</span> <span class="p">[]</span>
    <span class="n">segments</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">segments</span><span class="p">)</span>
    <span class="n">current_group</span> <span class="o">=</span> <span class="p">[</span><span class="n">segments</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span>
    <span class="n">groups</span> <span class="o">=</span> <span class="p">[</span><span class="n">current_group</span><span class="p">]</span>
    <span class="k">for</span> <span class="n">seg</span> <span class="ow">in</span> <span class="n">segments</span><span class="p">[</span><span class="mi">1</span><span class="p">:]:</span>
        <span class="n">gap_small_enough</span> <span class="o">=</span> <span class="p">((</span><span class="n">max_start_gap</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">)</span> <span class="ow">or</span>
                            <span class="p">(</span><span class="n">seg</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">-</span> <span class="n">current_group</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span> <span class="o">&lt;</span> <span class="n">max_start_gap</span><span class="p">))</span>
        <span class="n">spread_small_enough</span> <span class="o">=</span> <span class="p">((</span><span class="n">max_start_spread</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">)</span> <span class="ow">or</span>
                               <span class="p">(</span><span class="n">seg</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">-</span> <span class="n">current_group</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span> <span class="o">&lt;</span>
                                   <span class="n">max_start_spread</span><span class="p">))</span>
        <span class="k">if</span> <span class="n">gap_small_enough</span> <span class="ow">and</span> <span class="n">spread_small_enough</span><span class="p">:</span>
            <span class="n">current_group</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">seg</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">current_group</span> <span class="o">=</span> <span class="p">[</span><span class="n">seg</span><span class="p">]</span>
            <span class="n">groups</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">current_group</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">groups</span></div>

<div class="viewcode-block" id="annotate_record"><a class="viewcode-back" href="../../../ref/biotools.html#dnachisel.biotools.biotools.annotate_record">[docs]</a><span class="k">def</span> <span class="nf">annotate_record</span><span class="p">(</span><span class="n">seqrecord</span><span class="p">,</span> <span class="n">location</span><span class="o">=</span><span class="s2">&quot;full&quot;</span><span class="p">,</span> <span class="n">feature_type</span><span class="o">=</span><span class="s2">&quot;misc_feature&quot;</span><span class="p">,</span>
                    <span class="n">margin</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="o">**</span><span class="n">qualifiers</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Add a feature to a Biopython SeqRecord.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>

<span class="sd">    seqrecord</span>
<span class="sd">      The biopython seqrecord to be annotated.</span>

<span class="sd">    location</span>
<span class="sd">      Either (start, end) or (start, end, strand). (strand defaults to +1)</span>

<span class="sd">    feature_type</span>
<span class="sd">      The type associated with the feature</span>

<span class="sd">    margin</span>
<span class="sd">      Number of extra bases added on each side of the given location.</span>

<span class="sd">    qualifiers</span>
<span class="sd">      Dictionnary that will be the Biopython feature&#39;s `qualifiers` attribute.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">location</span> <span class="o">==</span> <span class="s2">&quot;full&quot;</span><span class="p">:</span>
        <span class="n">location</span> <span class="o">=</span> <span class="p">(</span><span class="n">margin</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">seqrecord</span><span class="p">)</span> <span class="o">-</span> <span class="n">margin</span><span class="p">)</span>

    <span class="n">strand</span> <span class="o">=</span> <span class="n">location</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span> <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">location</span><span class="p">)</span> <span class="o">==</span> <span class="mi">3</span> <span class="k">else</span> <span class="mi">1</span>
    <span class="n">seqrecord</span><span class="o">.</span><span class="n">features</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
        <span class="n">SeqFeature</span><span class="p">(</span>
            <span class="n">FeatureLocation</span><span class="p">(</span><span class="n">location</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">location</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">strand</span><span class="p">),</span>
            <span class="n">qualifiers</span><span class="o">=</span><span class="n">qualifiers</span><span class="p">,</span>
            <span class="nb">type</span><span class="o">=</span><span class="n">feature_type</span>
        <span class="p">)</span>
    <span class="p">)</span></div>
</pre></div>

           </div>
           <div class="articleComments">
            
           </div>
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2017, Edinburgh Genome Foundry.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'../../../',
            VERSION:'0.1.0',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true,
            SOURCELINK_SUFFIX: '.txt'
        };
    </script>
      <script type="text/javascript" src="../../../_static/jquery.js"></script>
      <script type="text/javascript" src="../../../_static/underscore.js"></script>
      <script type="text/javascript" src="../../../_static/doctools.js"></script>
      <script type="text/javascript" src="https://cdn.rawgit.com/knsv/mermaid/7.0.0/dist/mermaid.min.js"></script>

  

  
  
    <script type="text/javascript" src="../../../_static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
   

</body>
</html>